// @effect-ts/system/Async/excl-forEach.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var excl_forEach_exports = {};
__export(excl_forEach_exports, {
  collectAll: () => collectAll,
  collectAllPar: () => collectAllPar,
  collectAllUnit: () => collectAllUnit,
  collectAllUnitPar: () => collectAllUnitPar,
  collectAllWith: () => collectAllWith,
  collectAllWithPar: () => collectAllWithPar,
  collectAllWithPar_: () => collectAllWithPar_,
  collectAllWith_: () => collectAllWith_,
  forEach: () => forEach,
  forEachPar: () => forEachPar,
  forEachPar_: () => forEachPar_,
  forEachUnit: () => forEachUnit,
  forEachUnitPar: () => forEachUnitPar,
  forEachUnitPar_: () => forEachUnitPar_,
  forEachUnit_: () => forEachUnit_,
  forEach_: () => forEach_
});
var Collect = __toESM(require("../Collections/Immutable/Chunk/api/collect.js"));
var Chunk = __toESM(require("../Collections/Immutable/Chunk/core.js"));
var import_Function = require("../Function/index.js");
var I = __toESM(require("../Iterable/index.js"));
var core = __toESM(require("./core.js"));
function forEach_(as, f) {
  return core.suspend(() => {
    const acc = [];
    return core.map_(forEachUnit_(as, (a) => core.map_(f(a), (b) => {
      acc.push(b);
    })), () => Chunk.from(acc));
  });
}
function forEach(f) {
  return (as) => forEach_(as, f);
}
function forEachUnitLoop(iterator, f) {
  const next = iterator.next();
  return next.done ? core.unit : core.chain_(f(next.value), () => forEachUnitLoop(iterator, f));
}
function forEachUnit_(as, f) {
  return core.suspend(() => forEachUnitLoop(as[Symbol.iterator](), f));
}
function forEachUnit(f) {
  return (as) => forEachUnit_(as, f);
}
function forEachUnitPar_(as, f) {
  return core.chain_(core.environment(), (env) => core.promise((onInterrupt) => {
    return new Promise((resolve, reject) => {
      const is = new core.InterruptionState();
      const promises = [];
      onInterrupt(() => {
        is.interrupt();
      });
      const interruptOnFailure = (ex) => {
        if (ex._tag === "Failure" && !is.interrupted) {
          is.interrupt();
          reject(ex.e);
        }
      };
      for (const a of as) {
        promises.push(core.runPromiseExitEnv(f(a), env, is).then(interruptOnFailure));
      }
      Promise.all(promises).then(() => {
        resolve();
      });
    });
  }, (e) => e));
}
function forEachUnitPar(f) {
  return (as) => forEachUnitPar_(as, f);
}
function forEachPar_(as, f) {
  return core.suspend(() => core.chain_(core.succeedWith(() => []), (array) => core.map_(forEachUnitPar_(I.map_(as, (a, n) => [a, n]), ([a, n]) => core.chain_(core.suspend(() => f(a)), (b) => core.succeedWith(() => {
    array[n] = b;
  }))), () => Chunk.from(array))));
}
function forEachPar(f) {
  return (as) => forEachPar_(as, f);
}
function collectAll(as) {
  return forEach_(as, import_Function.identity);
}
function collectAllPar(as) {
  return forEachPar_(as, import_Function.identity);
}
function collectAllUnit(as) {
  return forEachUnit_(as, import_Function.identity);
}
function collectAllUnitPar(as) {
  return forEachUnitPar_(as, import_Function.identity);
}
function collectAllWith_(as, pf) {
  return core.map_(collectAll(as), Collect.collect(pf));
}
function collectAllWith(pf) {
  return (as) => collectAllWith_(as, pf);
}
function collectAllWithPar_(as, pf) {
  return core.map_(collectAllPar(as), Collect.collect(pf));
}
function collectAllWithPar(pf) {
  return (as) => collectAllWithPar_(as, pf);
}
module.exports = __toCommonJS(excl_forEach_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  collectAll,
  collectAllPar,
  collectAllUnit,
  collectAllUnitPar,
  collectAllWith,
  collectAllWithPar,
  collectAllWithPar_,
  collectAllWith_,
  forEach,
  forEachPar,
  forEachPar_,
  forEachUnit,
  forEachUnitPar,
  forEachUnitPar_,
  forEachUnit_,
  forEach_
});
