// @effect-ts/system/Pool/Attempted.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var Attempted_exports = {};
__export(Attempted_exports, {
  Attempted: () => Attempted,
  forEachUnit: () => forEachUnit,
  forEachUnit_: () => forEachUnit_,
  isFailure: () => isFailure,
  toManaged: () => toManaged
});
var T = __toESM(require("../Effect/index.js"));
var Ex = __toESM(require("../Exit/index.js"));
class Attempted {
  constructor(result, finalizer) {
    this.result = result;
    this.finalizer = finalizer;
  }
}
T._E, T._A;
function isFailure(self) {
  return self.result._tag === "Failure";
}
function forEachUnit_(self, f) {
  return Ex.foldM_(self.result, (_) => T.unit, (a) => f(a));
}
function forEachUnit(f) {
  return (self) => forEachUnit_(self, f);
}
function toManaged(self) {
  return T.toManaged(T.done(self.result));
}
module.exports = __toCommonJS(Attempted_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Attempted,
  forEachUnit,
  forEachUnit_,
  isFailure,
  toManaged
});
