// @effect-ts/core/Effect/instances.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var instances_exports = {};
__export(instances_exports, {
  Access: () => Access,
  Any: () => Any,
  Applicative: () => Applicative,
  AssociativeBoth: () => AssociativeBoth,
  AssociativeEither: () => AssociativeEither,
  AssociativeFlatten: () => AssociativeFlatten,
  Category: () => Category,
  Covariant: () => Covariant,
  EffectURI: () => import_Modules.EffectURI,
  Fail: () => Fail,
  IdentityBoth: () => IdentityBoth,
  IdentityFlatten: () => IdentityFlatten,
  Monad: () => Monad,
  Provide: () => Provide,
  Run: () => Run,
  getIdentity: () => getIdentity,
  getIdentityPar: () => getIdentityPar,
  getValidationApplicative: () => getValidationApplicative,
  match: () => match,
  matchIn: () => matchIn,
  matchMorph: () => matchMorph,
  matchTag: () => matchTag,
  matchTagIn: () => matchTagIn
});
var import_Operator = require("../Operator/index.js");
var T = __toESM(require("@effect-ts/system/Effect"));
var I = __toESM(require("../Identity/index.js"));
var DSL = __toESM(require("../Prelude/DSL/index.js"));
var P = __toESM(require("../Prelude/index.js"));
__reExport(instances_exports, require("@effect-ts/system/Effect"));
var import_Modules = require("../Modules/index.js");
const Any = {
  any: () => T.succeed({})
};
const AssociativeEither = {
  orElseEither: T.orElseEither
};
const AssociativeFlatten = {
  flatten: T.flatten
};
const AssociativeBoth = {
  both: T.zip
};
const Covariant = {
  map: T.map
};
const IdentityFlatten = __spreadValues(__spreadValues({}, Any), AssociativeFlatten);
const IdentityBoth = __spreadValues(__spreadValues({}, Any), AssociativeBoth);
const Monad = __spreadValues(__spreadValues({}, IdentityFlatten), Covariant);
const Applicative = __spreadValues(__spreadValues({}, Covariant), IdentityBoth);
const Fail = {
  fail: T.fail
};
const Run = {
  either: T.either
};
const Access = {
  access: T.access
};
const Provide = {
  provide: T.provideAll
};
const getValidationApplicative = /* @__PURE__ */ DSL.getValidationF(__spreadValues(__spreadValues(__spreadValues(__spreadValues({}, Monad), Run), Fail), Applicative));
const Category = {
  id: T.environment,
  compose: T.compose
};
const {
  match,
  matchIn,
  matchMorph,
  matchTag,
  matchTagIn
} = /* @__PURE__ */ DSL.matchers(Covariant);
function getIdentity(Id) {
  return () => I.makeIdentity(T.succeed(Id.identity), (x, y) => T.zipWith_(x, y, Id.combine));
}
function getIdentityPar(Id) {
  return () => I.makeIdentity(T.succeed(Id.identity), (x, y) => T.zipWithPar_(x, y, Id.combine));
}
module.exports = __toCommonJS(instances_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Access,
  Any,
  Applicative,
  AssociativeBoth,
  AssociativeEither,
  AssociativeFlatten,
  Category,
  Covariant,
  EffectURI,
  Fail,
  IdentityBoth,
  IdentityFlatten,
  Monad,
  Provide,
  Run,
  getIdentity,
  getIdentityPar,
  getValidationApplicative,
  match,
  matchIn,
  matchMorph,
  matchTag,
  matchTagIn
});
