// @effect-ts/system/Testing/TestRunner/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var TestRunner_exports = {};
__export(TestRunner_exports, {
  TestRunner: () => TestRunner,
  defaultTestRunner: () => defaultTestRunner
});
var Clock = __toESM(require("../../Clock/index.js"));
var T = __toESM(require("../../Effect/index.js"));
var import_Function = require("../../Function/index.js");
var L = __toESM(require("../../Layer/index.js"));
var Annotations = __toESM(require("../Annotations/index.js"));
var TestAnnotationRenderer = __toESM(require("../TestAnnotationRenderer/index.js"));
var import_TestClock = require("../TestClock/index.js");
var import_TestExecutor = require("../TestExecutor/index.js");
var import_TestLogger = require("../TestLogger/index.js");
var import_TestReporter = require("../TestReporter/index.js");
const defaultClock = /* @__PURE__ */ L.fromFunction(Clock.HasClock)(() => new Clock.LiveClock());
const defaultLayer = /* @__PURE__ */ defaultClock["+++"](import_TestLogger.FromConsole);
class TestRunner {
  constructor(executor, platform = T.defaultPlatform, reporter = (0, import_TestReporter.DefaultTestReporter)(TestAnnotationRenderer.standard), bootstrap = defaultLayer) {
    this.executor = executor;
    this.platform = platform;
    this.reporter = reporter;
    this.bootstrap = bootstrap;
    this.runtime = new T.CustomRuntime(void 0, this.platform);
    this.run = (spec) => T.chain_(T.timed(this.executor.run(spec, T.parallelN(4))), ({
      tuple: [d, e]
    }) => T.as_(this.reporter((0, import_TestClock.Duration)(d), e), e));
  }
}
const defaultTestRunner = /* @__PURE__ */ new TestRunner(/* @__PURE__ */ (0, import_TestExecutor.defaultExecutor)(/* @__PURE__ */ Annotations.live["+++"](/* @__PURE__ */ L.succeed(T.defaultEnv))));
module.exports = __toCommonJS(TestRunner_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TestRunner,
  defaultTestRunner
});
