// @effect-ts/core/Associative/common.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all2) => {
  for (var name in all2)
    __defProp(target, name, { get: all2[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var common_exports = {};
__export(common_exports, {
  all: () => all,
  any: () => any,
  product: () => product,
  string: () => string,
  sum: () => sum,
  void: () => void_
});
var import_makeAssociative = require("./makeAssociative.js");
__reExport(common_exports, require("./definition.js"));
const all = /* @__PURE__ */ (0, import_makeAssociative.makeAssociative)((x, y) => x && y);
const any = /* @__PURE__ */ (0, import_makeAssociative.makeAssociative)((x, y) => x || y);
const sum = /* @__PURE__ */ (0, import_makeAssociative.makeAssociative)((x, y) => x + y);
const product = /* @__PURE__ */ (0, import_makeAssociative.makeAssociative)((x, y) => x * y);
const string = /* @__PURE__ */ (0, import_makeAssociative.makeAssociative)((x, y) => x + y);
const void_ = /* @__PURE__ */ (0, import_makeAssociative.makeAssociative)(() => void 0);
module.exports = __toCommonJS(common_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  all,
  any,
  product,
  string,
  sum,
  void: null
});
