// @effect-ts/core/Reader/operations.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var operations_exports = {};
__export(operations_exports, {
  access: () => access,
  chain: () => chain,
  environment: () => environment,
  map: () => map,
  provideSome: () => provideSome,
  run: () => run,
  runEnv: () => runEnv,
  succeed: () => succeed,
  sync: () => sync,
  tap: () => tap,
  zip: () => zip
});
var Tp = __toESM(require("@effect-ts/system/Collections/Immutable/Tuple"));
var import_Function = require("../Function/index.js");
const sync = import_Function.identity;
const environment = () => import_Function.identity;
const access = import_Function.identity;
const provideSome = (f) => (ma) => (r) => ma(f(r));
const zip = (fb) => (fa) => (r) => Tp.tuple(fa(r), fb(r));
const chain = (f) => (fa) => (r) => f(fa(r))(r);
const map = (f) => (fa) => (r) => f(fa(r));
const succeed = (a) => () => a;
const run = (self) => self({});
const runEnv = (r) => (self) => self(r);
const tap = (f) => (fa) => (r) => {
  const x = fa(r);
  f(x)(r);
  return x;
};
module.exports = __toCommonJS(operations_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  access,
  chain,
  environment,
  map,
  provideSome,
  run,
  runEnv,
  succeed,
  sync,
  tap,
  zip
});
