// @effect-ts/system/FiberRef/modify.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var modify_exports = {};
__export(modify_exports, {
  modify: () => modify,
  modify_: () => modify_
});
var Tp = __toESM(require("../Collections/Immutable/Tuple/index.js"));
var E = __toESM(require("../Either/index.js"));
var import_Function = require("../Function/index.js");
var import_Utils = require("../Utils/index.js");
var T = __toESM(require("./excl-effect.js"));
var import_fiberRef = require("./fiberRef.js");
function modify(f) {
  return (fiberRef) => modify_(fiberRef, f);
}
function modify_(fiberRef, f) {
  return (0, import_Utils.matchTag)({
    Runtime: (self) => self.modify(f),
    Derived: (self) => self.use((value, getEither, setEither) => T.absolve(value.modify((s) => E.fold_(getEither(s), (e) => Tp.tuple(E.left(e), s), (a1) => (({
      tuple: [b, a2]
    }) => E.fold_(setEither(a2), (e) => Tp.tuple(E.left(e), s), (s2) => Tp.tuple(E.right(b), s2)))(f(a1)))))),
    DerivedAll: (self) => self.use((value, getEither, setEither) => T.absolve(value.modify((s) => E.fold_(getEither(s), (e) => Tp.tuple(E.left(e), s), (a1) => (({
      tuple: [b, a2]
    }) => E.fold_(setEither(a2)(s), (e) => Tp.tuple(E.left(e), s), (s2) => Tp.tuple(E.right(b), s2)))(f(a1))))))
  })(fiberRef);
}
module.exports = __toCommonJS(modify_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  modify,
  modify_
});
