// @effect-ts/system/Testing/TestFailure/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var TestFailure_exports = {};
__export(TestFailure_exports, {
  Assertion: () => Assertion,
  AssertionTypeId: () => AssertionTypeId,
  Runtime: () => Runtime,
  RuntimeTypeId: () => RuntimeTypeId,
  assertion: () => assertion,
  die: () => die,
  fail: () => fail,
  halt: () => halt
});
var C = __toESM(require("../../Cause/index.js"));
var T = __toESM(require("../../Effect/index.js"));
const AssertionTypeId = /* @__PURE__ */ Symbol();
class Assertion {
  constructor(result) {
    this.result = result;
    this._typeId = AssertionTypeId;
  }
}
const RuntimeTypeId = /* @__PURE__ */ Symbol();
class Runtime {
  constructor(cause) {
    this.cause = cause;
    this._typeId = RuntimeTypeId;
  }
}
T._E;
function assertion(result) {
  return new Assertion(result);
}
function die(e) {
  return halt(C.die(e));
}
function fail(e) {
  return halt(C.fail(e));
}
function halt(cause) {
  return new Runtime(cause);
}
module.exports = __toCommonJS(TestFailure_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Assertion,
  AssertionTypeId,
  Runtime,
  RuntimeTypeId,
  assertion,
  die,
  fail,
  halt
});
