// @effect-ts/core/Identity/common.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all2) => {
  for (var name in all2)
    __defProp(target, name, { get: all2[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var common_exports = {};
__export(common_exports, {
  all: () => all,
  any: () => any,
  fromAssociative: () => fromAssociative,
  product: () => product,
  string: () => string,
  sum: () => sum,
  void: () => void_
});
var A = __toESM(require("../Associative/common.js"));
var import_makeIdentity = require("./makeIdentity.js");
function fromAssociative(A2) {
  return (identity) => (0, import_makeIdentity.makeIdentity)(identity, A2.combine);
}
const all = /* @__PURE__ */ (0, import_makeIdentity.makeIdentity)(true, A.all.combine);
const any = /* @__PURE__ */ fromAssociative(A.any)(false);
const product = /* @__PURE__ */ fromAssociative(A.product)(1);
const string = /* @__PURE__ */ fromAssociative(A.string)("");
const sum = /* @__PURE__ */ fromAssociative(A.sum)(0);
const void_ = /* @__PURE__ */ fromAssociative(A.void)(void 0);
module.exports = __toCommonJS(common_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  all,
  any,
  fromAssociative,
  product,
  string,
  sum,
  void: null
});
