// @effect-ts/system/Random/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var Random_exports = {};
__export(Random_exports, {
  HasRandom: () => HasRandom,
  LiveRandom: () => LiveRandom,
  Random: () => Random,
  RandomId: () => RandomId,
  defaultRandom: () => defaultRandom,
  next: () => next,
  nextBoolean: () => nextBoolean,
  nextInt: () => nextInt,
  nextIntBetween: () => nextIntBetween,
  nextRange: () => nextRange,
  withSeed: () => withSeed
});
var import_Operator = require("../Operator/index.js");
var import_core = require("../Effect/core.js");
var import_has = require("../Effect/has.js");
var import_Has = require("../Has/index.js");
var import_PCG = require("./PCG/index.js");
const RandomId = /* @__PURE__ */ Symbol.for("@effect-ts/system/Random");
class Random {
  constructor() {
    this.serviceId = RandomId;
  }
}
class LiveRandom extends Random {
  constructor(seed) {
    super();
    this.next = (0, import_core.succeedWith)(() => this.PRNG.number());
    this.nextBoolean = (0, import_core.chain_)(this.next, (n) => (0, import_core.succeedWith)(() => n > 0.5));
    this.nextInt = (0, import_core.succeedWith)(() => this.PRNG.integer(0));
    this.nextRange = (low, high) => (0, import_core.chain_)(this.next, (n) => (0, import_core.succeedWith)(() => (high - low) * n + low));
    this.nextIntBetween = (low, high) => (0, import_core.succeedWith)(() => this.PRNG.integer(1 + high - low) + low);
    this.PRNG = new import_PCG.PCGRandom(seed);
  }
}
const defaultRandom = /* @__PURE__ */ new LiveRandom(/* @__PURE__ */ Math.random() * 4294967296 >>> 0);
const HasRandom = /* @__PURE__ */ (0, import_Has.tag)(RandomId);
const next = /* @__PURE__ */ (0, import_has.accessServiceM)(HasRandom)((_) => _.next);
const nextBoolean = /* @__PURE__ */ (0, import_has.accessServiceM)(HasRandom)((_) => _.nextBoolean);
const nextIntBetween = (low, high) => (0, import_has.accessServiceM)(HasRandom)((_) => _.nextIntBetween(low, high));
const nextInt = /* @__PURE__ */ (0, import_has.accessServiceM)(HasRandom)((_) => _.nextInt);
const nextRange = (low, high) => (0, import_has.accessServiceM)(HasRandom)((_) => _.nextRange(low, high));
const withSeed = (seed) => (0, import_has.replaceService)(HasRandom, () => new LiveRandom(seed));
module.exports = __toCommonJS(Random_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  HasRandom,
  LiveRandom,
  Random,
  RandomId,
  defaultRandom,
  next,
  nextBoolean,
  nextInt,
  nextIntBetween,
  nextRange,
  withSeed
});
