// @effect-ts/system/Testing/Render/api.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var api_exports = {};
__export(api_exports, {
  className: () => className,
  field: () => field,
  function_: () => function_,
  infix: () => infix,
  param: () => param,
  quoted: () => quoted
});
var AssertionM = __toESM(require("../AssertionM/AssertionM.js"));
var R = __toESM(require("./definition.js"));
function className(cons) {
  return cons.prototype.constructor.name;
}
function field(name) {
  return `_.${name}`;
}
function function_(name, paramLists) {
  return new R.Function_(name, paramLists);
}
function infix(left, op, right) {
  return new R.Infix(left, op, right);
}
function param(value) {
  if (AssertionM.isAssertionM(value)) {
    return new R.AssertionM(value);
  }
  return new R.Value(value);
}
function quoted(str) {
  return `"${str}"`;
}
module.exports = __toCommonJS(api_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  className,
  field,
  function_,
  infix,
  param,
  quoted
});
