// @effect-ts/core/Sync/_internal/dsl.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var dsl_exports = {};
__export(dsl_exports, {
  getValidationApplicative: () => getValidationApplicative,
  if: () => branch,
  if_: () => branch_,
  match: () => match,
  matchIn: () => matchIn,
  matchMorph: () => matchMorph,
  matchTag: () => matchTag,
  matchTagIn: () => matchTagIn,
  struct: () => struct,
  tuple: () => tuple
});
var P = __toESM(require("../../Prelude/index.js"));
var import_instances = require("./instances.js");
const tuple = /* @__PURE__ */ P.tupleF(import_instances.Applicative);
const struct = /* @__PURE__ */ P.structF(import_instances.Applicative);
const getValidationApplicative = /* @__PURE__ */ P.getValidationF(__spreadValues(__spreadValues(__spreadValues(__spreadValues({}, import_instances.Applicative), import_instances.Fail), import_instances.Run), import_instances.Monad));
const {
  match,
  matchIn,
  matchMorph,
  matchTag,
  matchTagIn
} = /* @__PURE__ */ P.matchers(import_instances.Covariant);
const branch = /* @__PURE__ */ P.conditionalF(import_instances.Covariant);
const branch_ = /* @__PURE__ */ P.conditionalF_(import_instances.Covariant);
module.exports = __toCommonJS(dsl_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getValidationApplicative,
  if: null,
  if_,
  match,
  matchIn,
  matchMorph,
  matchTag,
  matchTagIn,
  struct,
  tuple
});
