// @effect-ts/core/Collections/Immutable/Chunk/instances.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var instances_exports = {};
__export(instances_exports, {
  Any: () => Any,
  Applicative: () => Applicative,
  ApplyZip: () => ApplyZip,
  AssociativeBothZip: () => AssociativeBothZip,
  AssociativeFlatten: () => AssociativeFlatten,
  BreadthFirstChainRec: () => BreadthFirstChainRec,
  Collection: () => Collection,
  Compact: () => Compact,
  Covariant: () => Covariant,
  DepthFirstChainRec: () => DepthFirstChainRec,
  Extend: () => Extend,
  Filter: () => Filter,
  FilterMap: () => FilterMap,
  FilterMapWithIndex: () => FilterMapWithIndex,
  FilterWithIndex: () => FilterWithIndex,
  Filterable: () => Filterable,
  FilterableWithIndex: () => FilterableWithIndex,
  FoldMap: () => FoldMap,
  FoldMapWithIndex: () => FoldMapWithIndex,
  Foldable: () => Foldable,
  FoldableWithIndex: () => FoldableWithIndex,
  ForEach: () => ForEach,
  ForEachWithIndex: () => ForEachWithIndex,
  Monad: () => Monad,
  Partition: () => Partition,
  PartitionMap: () => PartitionMap,
  PartitionMapWithIndex: () => PartitionMapWithIndex,
  PartitionWithIndex: () => PartitionWithIndex,
  Reduce: () => Reduce,
  ReduceRight: () => ReduceRight,
  ReduceRightWithIndex: () => ReduceRightWithIndex,
  ReduceWithIndex: () => ReduceWithIndex,
  Separate: () => Separate,
  Wiltable: () => Wiltable,
  WiltableWithIndex: () => WiltableWithIndex,
  Witherable: () => Witherable,
  WitherableWithIndex: () => WitherableWithIndex,
  breadthFirstChainRec: () => breadthFirstChainRec,
  depthFirstChainRec: () => depthFirstChainRec
});
var O = __toESM(require("../../../Option/index.js"));
var import_Prelude = require("../../../Prelude/index.js");
var P = __toESM(require("../../../Prelude/index.js"));
var Chunk = __toESM(require("./operations.js"));
const Collection = {
  builder: Chunk.builder
};
const Any = {
  any: () => Chunk.single({})
};
const AssociativeBothZip = {
  both: Chunk.zip
};
const AssociativeFlatten = {
  flatten: Chunk.flatten
};
const Covariant = {
  map: Chunk.map
};
const ApplyZip = __spreadValues(__spreadValues({}, Covariant), AssociativeBothZip);
const Monad = __spreadValues(__spreadValues(__spreadValues({}, Any), Covariant), AssociativeFlatten);
const Applicative = /* @__PURE__ */ (0, import_Prelude.getApplicativeF)(Monad);
const ForEach = {
  map: Chunk.map,
  forEachF: Chunk.forEachF
};
const ForEachWithIndex = {
  map: Chunk.map,
  forEachWithIndexF: Chunk.forEachWithIndexF
};
const Wiltable = {
  separateF: Chunk.separateF
};
const WiltableWithIndex = {
  separateWithIndexF: Chunk.separateWithIndexF
};
const Witherable = {
  compactF: Chunk.compactF
};
const WitherableWithIndex = {
  compactWithIndexF: Chunk.compactWithIndexF
};
const Compact = {
  compact: Chunk.compact
};
const Separate = {
  separate: Chunk.separate
};
const Extend = {
  extend: (f) => (fa) => Chunk.single(f(fa))
};
const Reduce = {
  reduce: Chunk.reduce
};
const ReduceWithIndex = {
  reduceWithIndex: (b, f) => (fa) => Chunk.reduce_(Chunk.zipWithIndex(fa), b, (b2, {
    tuple: [a, i]
  }) => f(i, b2, a))
};
const ReduceRightWithIndex = {
  reduceRightWithIndex: (b, f) => (fa) => Chunk.reduceRight_(Chunk.zipWithIndex(fa), b, ({
    tuple: [a, i]
  }, b2) => f(i, a, b2))
};
const ReduceRight = {
  reduceRight: Chunk.reduceRight
};
const FoldMap = {
  foldMap: Chunk.foldMap
};
const FoldMapWithIndex = {
  foldMapWithIndex: Chunk.foldMapWithIndex
};
const Foldable = __spreadValues(__spreadValues(__spreadValues({}, FoldMap), Reduce), ReduceRight);
const FoldableWithIndex = __spreadValues(__spreadValues(__spreadValues({}, FoldMapWithIndex), ReduceWithIndex), ReduceRightWithIndex);
const Filter = {
  filter: Chunk.filter
};
const FilterWithIndex = {
  filterWithIndex: (predicate) => (fa) => Chunk.collect_(Chunk.zipWithIndex(fa), ({
    tuple: [a, i]
  }) => predicate(i, a) ? O.some(a) : O.none)
};
const FilterMap = {
  filterMap: Chunk.collect
};
const FilterMapWithIndex = {
  filterMapWithIndex: (f) => (fa) => Chunk.collect_(Chunk.zipWithIndex(fa), ({
    tuple: [a, i]
  }) => f(i, a))
};
const Partition = {
  partition: Chunk.partition
};
const PartitionWithIndex = {
  partitionWithIndex: Chunk.partitionWithIndex
};
const PartitionMap = {
  partitionMap: Chunk.partitionMap
};
const PartitionMapWithIndex = {
  partitionMapWithIndex: Chunk.partitionMapWithIndex
};
const Filterable = __spreadValues(__spreadValues(__spreadValues(__spreadValues({}, Filter), FilterMap), Partition), PartitionMap);
const FilterableWithIndex = __spreadValues(__spreadValues(__spreadValues(__spreadValues({}, FilterWithIndex), FilterMapWithIndex), PartitionWithIndex), PartitionMapWithIndex);
const DepthFirstChainRec = {
  chainRec: (f) => (a) => {
    let todo = f(a);
    let result = Chunk.empty();
    while (Chunk.size(todo) > 0) {
      const e = Chunk.unsafeHead(todo);
      todo = Chunk.unsafeTail(todo);
      if (e._tag === "Left") {
        todo = Chunk.concat_(f(e.left), todo);
      } else {
        result = Chunk.append_(result, e.right);
      }
    }
    return result;
  }
};
const depthFirstChainRec = DepthFirstChainRec.chainRec;
const BreadthFirstChainRec = {
  chainRec: (f) => (a) => {
    let todo = f(a);
    let result = Chunk.empty();
    while (Chunk.size(todo) > 0) {
      const e = Chunk.unsafeHead(todo);
      todo = Chunk.unsafeTail(todo);
      if (e._tag === "Left") {
        todo = Chunk.concat_(todo, f(e.left));
      } else {
        result = Chunk.append_(result, e.right);
      }
    }
    return result;
  }
};
const breadthFirstChainRec = BreadthFirstChainRec.chainRec;
module.exports = __toCommonJS(instances_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Any,
  Applicative,
  ApplyZip,
  AssociativeBothZip,
  AssociativeFlatten,
  BreadthFirstChainRec,
  Collection,
  Compact,
  Covariant,
  DepthFirstChainRec,
  Extend,
  Filter,
  FilterMap,
  FilterMapWithIndex,
  FilterWithIndex,
  Filterable,
  FilterableWithIndex,
  FoldMap,
  FoldMapWithIndex,
  Foldable,
  FoldableWithIndex,
  ForEach,
  ForEachWithIndex,
  Monad,
  Partition,
  PartitionMap,
  PartitionMapWithIndex,
  PartitionWithIndex,
  Reduce,
  ReduceRight,
  ReduceRightWithIndex,
  ReduceWithIndex,
  Separate,
  Wiltable,
  WiltableWithIndex,
  Witherable,
  WitherableWithIndex,
  breadthFirstChainRec,
  depthFirstChainRec
});
