// @effect-ts/core/Collections/Immutable/Array/interop.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var interop_exports = {};
__export(interop_exports, {
  mapEffect: () => mapEffect,
  mapEffectPar: () => mapEffectPar,
  mapEffectParN: () => mapEffectParN,
  mapEffectParN_: () => mapEffectParN_,
  mapEffectPar_: () => mapEffectPar_,
  mapEffect_: () => mapEffect_,
  mapSync: () => mapSync,
  mapSync_: () => mapSync_
});
var T = __toESM(require("@effect-ts/system/Effect"));
var S = __toESM(require("@effect-ts/system/Sync"));
var A = __toESM(require("./operations.js"));
function mapEffect_(self, f) {
  return T.map_(T.forEach_(self, f), A.from);
}
function mapEffect(f) {
  return (self) => mapEffect_(self, f);
}
function mapEffectPar_(self, f) {
  return T.map_(T.forEachPar_(self, f), A.from);
}
function mapEffectPar(f) {
  return (self) => mapEffectPar_(self, f);
}
function mapEffectParN_(self, n, f) {
  return T.map_(T.forEachParN_(self, n, f), A.from);
}
function mapEffectParN(n, f) {
  return (self) => mapEffectParN_(self, n, f);
}
function mapSync_(self, f) {
  return S.map_(S.forEach_(self, f), A.from);
}
function mapSync(f) {
  return (self) => mapSync_(self, f);
}
module.exports = __toCommonJS(interop_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  mapEffect,
  mapEffectPar,
  mapEffectParN,
  mapEffectParN_,
  mapEffectPar_,
  mapEffect_,
  mapSync,
  mapSync_
});
