// @effect-ts/system/Experimental/Stream/Pull/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var Pull_exports = {};
__export(Pull_exports, {
  emit: () => emit,
  emitChunk: () => emitChunk,
  empty: () => empty,
  end: () => end,
  fail: () => fail,
  failCause: () => failCause,
  fromQueue: () => fromQueue
});
var A = __toESM(require("../../../Collections/Immutable/Chunk/index.js"));
var T = __toESM(require("../../../Effect/index.js"));
var O = __toESM(require("../../../Option/index.js"));
var Q = __toESM(require("../../../Queue/index.js"));
var Take = __toESM(require("../Take/index.js"));
function emit(a) {
  return T.succeed(A.single(a));
}
function emitChunk(as) {
  return T.succeed(as);
}
function fromQueue(d) {
  return T.chain_(Q.take(d), (_) => Take.done(_));
}
function fail(e) {
  return T.fail(O.some(e));
}
function failCause(c) {
  return T.mapError_(T.halt(c), O.some);
}
function empty() {
  return T.succeed(A.empty());
}
const end = /* @__PURE__ */ T.fail(O.none);
module.exports = __toCommonJS(Pull_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  emit,
  emitChunk,
  empty,
  end,
  fail,
  failCause,
  fromQueue
});
