// @effect-ts/system/Queue/unsafe.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var unsafe_exports = {};
__export(unsafe_exports, {
  unsafeCreateQueue: () => unsafeCreateQueue,
  unsafeCreateQueue_: () => unsafeCreateQueue_,
  unsafeMakeBounded: () => unsafeMakeBounded,
  unsafeMakeDropping: () => unsafeMakeDropping,
  unsafeMakeSliding: () => unsafeMakeSliding,
  unsafeMakeUnbounded: () => unsafeMakeUnbounded
});
var import_id = require("../Fiber/id.js");
var import_AtomicBoolean = require("../Support/AtomicBoolean/index.js");
var import_MutableQueue = require("../Support/MutableQueue/index.js");
var import_api = require("./api.js");
var import_core = require("./core.js");
var P = __toESM(require("./promise.js"));
function unsafeCreateQueue(strategy) {
  return (queue) => unsafeCreateQueue_(queue, strategy);
}
function unsafeCreateQueue_(queue, strategy) {
  return (0, import_api.unsafeCreate)(queue, new import_MutableQueue.Unbounded(), P.unsafeMake(import_id.None), new import_AtomicBoolean.AtomicBoolean(false), strategy);
}
function unsafeMakeSliding(capacity) {
  return unsafeCreateQueue_(new import_MutableQueue.Bounded(capacity), new import_core.SlidingStrategy());
}
function unsafeMakeUnbounded() {
  return unsafeCreateQueue_(new import_MutableQueue.Unbounded(), new import_core.DroppingStrategy());
}
function unsafeMakeDropping(capacity) {
  return unsafeCreateQueue_(new import_MutableQueue.Bounded(capacity), new import_core.DroppingStrategy());
}
function unsafeMakeBounded(capacity) {
  return unsafeCreateQueue_(new import_MutableQueue.Bounded(capacity), new import_api.BackPressureStrategy());
}
module.exports = __toCommonJS(unsafe_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  unsafeCreateQueue,
  unsafeCreateQueue_,
  unsafeMakeBounded,
  unsafeMakeDropping,
  unsafeMakeSliding,
  unsafeMakeUnbounded
});
