// @effect-ts/system/Experimental/Stream/_internal/SinkEndReason.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var SinkEndReason_exports = {};
__export(SinkEndReason_exports, {
  ScheduleEnd: () => ScheduleEnd,
  ScheduleEndTypeId: () => ScheduleEndTypeId,
  ScheduleTimeout: () => ScheduleTimeout,
  ScheduleTimeoutTypeId: () => ScheduleTimeoutTypeId,
  SinkEnd: () => SinkEnd,
  SinkEndReasonTypeId: () => SinkEndReasonTypeId,
  SinkEndTypeId: () => SinkEndTypeId,
  UpstreamEnd: () => UpstreamEnd,
  UpstreamEndTypeId: () => UpstreamEndTypeId
});
const SinkEndReasonTypeId = /* @__PURE__ */ Symbol();
const SinkEndTypeId = /* @__PURE__ */ Symbol();
class SinkEnd {
  constructor() {
    this._sinkEndReasonTypeId = SinkEndReasonTypeId;
    this._typeId = SinkEndTypeId;
  }
}
const ScheduleTimeoutTypeId = /* @__PURE__ */ Symbol();
class ScheduleTimeout {
  constructor() {
    this._sinkEndReasonTypeId = SinkEndReasonTypeId;
    this._typeId = ScheduleTimeoutTypeId;
  }
}
const ScheduleEndTypeId = /* @__PURE__ */ Symbol();
class ScheduleEnd {
  constructor(c) {
    this.c = c;
    this._sinkEndReasonTypeId = SinkEndReasonTypeId;
    this._typeId = ScheduleEndTypeId;
  }
}
const UpstreamEndTypeId = /* @__PURE__ */ Symbol();
class UpstreamEnd {
  constructor() {
    this._sinkEndReasonTypeId = SinkEndReasonTypeId;
    this._typeId = UpstreamEndTypeId;
  }
}
module.exports = __toCommonJS(SinkEndReason_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ScheduleEnd,
  ScheduleEndTypeId,
  ScheduleTimeout,
  ScheduleTimeoutTypeId,
  SinkEnd,
  SinkEndReasonTypeId,
  SinkEndTypeId,
  UpstreamEnd,
  UpstreamEndTypeId
});
