// @effect-ts/system/Exit/_internal/cause.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var cause_exports = {};
__export(cause_exports, {
  both: () => import_cause.both,
  die: () => import_cause.die,
  empty: () => import_cause.empty,
  fail: () => import_cause.fail,
  interrupt: () => import_cause.interrupt,
  then: () => import_cause.then
});
var import_cause = require("../../Cause/cause.js");
module.exports = __toCommonJS(cause_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  both,
  die,
  empty,
  fail,
  interrupt,
  then
});
