// @effect-ts/system/Hub/_internal/unsafe.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var unsafe_exports = {};
__export(unsafe_exports, {
  unsafeCompletePromise: () => unsafeCompletePromise,
  unsafeOfferAll: () => unsafeOfferAll,
  unsafePollAllQueue: () => unsafePollAllQueue,
  unsafePollAllSubscription: () => unsafePollAllSubscription,
  unsafePollN: () => unsafePollN,
  unsafePublishAll: () => unsafePublishAll,
  unsafeRemove: () => unsafeRemove
});
var import_Operator = require("../../Operator/index.js");
var Chunk = __toESM(require("../../Collections/Immutable/Chunk/index.js"));
var T = __toESM(require("../../Effect/index.js"));
var P = __toESM(require("../../Promise/index.js"));
function unsafeCompletePromise(promise, a) {
  P.unsafeDone(T.succeed(a))(promise);
}
function unsafeOfferAll(queue, as) {
  return queue.offerAll(as);
}
function unsafePollAllQueue(queue) {
  return queue.pollUpTo(Number.MAX_SAFE_INTEGER);
}
function unsafePollAllSubscription(subscription) {
  return subscription.pollUpTo(Number.MAX_SAFE_INTEGER);
}
function unsafePollN(subscription, max) {
  return subscription.pollUpTo(max);
}
function unsafePublishAll(hub, as) {
  return hub.publishAll(as);
}
function unsafeRemove(queue, a) {
  unsafeOfferAll(queue, Chunk.filter_(unsafePollAllQueue(queue), (_) => _ !== a));
}
module.exports = __toCommonJS(unsafe_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  unsafeCompletePromise,
  unsafeOfferAll,
  unsafePollAllQueue,
  unsafePollAllSubscription,
  unsafePollN,
  unsafePublishAll,
  unsafeRemove
});
