// @effect-ts/system/Utils/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var Utils_exports = {};
__export(Utils_exports, {
  intersect: () => intersect,
  isAdtElement: () => isAdtElement,
  isEither: () => isEither,
  isGenericAdtElement: () => isGenericAdtElement,
  isOption: () => isOption,
  isSync: () => isSync,
  isTag: () => isTag,
  matchTag: () => matchTag,
  matchTagFor: () => matchTagFor,
  matchTag_: () => matchTag_,
  onAdtElement: () => onAdtElement,
  onGenericAdtElement: () => onGenericAdtElement,
  pattern: () => pattern,
  patternFor: () => patternFor,
  pattern_: () => pattern_
});
var import_Operator = require("../Operator/index.js");
var import_Option = require("../Option/index.js");
__reExport(Utils_exports, require("./tool.js"));
__reExport(Utils_exports, require("./lazy.js"));
__reExport(Utils_exports, require("./union.js"));
__reExport(Utils_exports, require("./equal.js"));
__reExport(Utils_exports, require("./unification.js"));
function intersect(...as) {
  return as.reduce((a, b) => __spreadValues(__spreadValues({}, a), b));
}
const pattern = (n) => (_, d) => (m) => {
  return _[m[n]] ? _[m[n]](m, m) : d(m, m);
};
const matchTag = /* @__PURE__ */ pattern("_tag");
const pattern_ = (n) => (m, _, d) => {
  return _[m[n]] ? _[m[n]](m, m) : d(m, m);
};
const matchTag_ = /* @__PURE__ */ pattern_("_tag");
const patternFor = (n) => () => (_, d) => (m) => {
  return _[m[n]] ? _[m[n]](m, m) : d(m, m);
};
const matchTagFor = /* @__PURE__ */ patternFor("_tag");
function isEither(u) {
  return typeof u === "object" && u != null && "_tag" in u && (u["_tag"] === "Left" || u["_tag"] === "Right");
}
function isOption(u) {
  return typeof u === "object" && u != null && "_tag" in u && (u["_tag"] === "Some" || u["_tag"] === "None");
}
function isTag(u) {
  return typeof u === "object" && u != null && "_tag" in u && u["_tag"] === "Tag";
}
function isSync(u) {
  return typeof u === "object" && u != null && "_tag" in u && u["_tag"] === "XPure";
}
function isAdtElement(tag) {
  return (adt) => adt["_tag"] === tag;
}
function isGenericAdtElement(_t) {
  return (tag) => (adt) => adt[_t] === tag;
}
function onAdtElement(tag, f) {
  return (adt) => {
    if (adt["_tag"] === tag) {
      return (0, import_Option.some)(f(adt));
    }
    return import_Option.none;
  };
}
function onGenericAdtElement(_t) {
  return (tag, f) => (adt) => {
    if (adt[_t] === tag) {
      return (0, import_Option.some)(f(adt));
    }
    return import_Option.none;
  };
}
module.exports = __toCommonJS(Utils_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  intersect,
  isAdtElement,
  isEither,
  isGenericAdtElement,
  isOption,
  isSync,
  isTag,
  matchTag,
  matchTagFor,
  matchTag_,
  onAdtElement,
  onGenericAdtElement,
  pattern,
  patternFor,
  pattern_
});
