// @effect-ts/system/Effect/bindAll.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var bindAll_exports = {};
__export(bindAll_exports, {
  bindAll: () => bindAll,
  bindAllPar: () => bindAllPar,
  bindAllParN: () => bindAllParN,
  bindAllParN_: () => bindAllParN_,
  bindAllPar_: () => bindAllPar_,
  bindAll_: () => bindAll_
});
var R = __toESM(require("../Collections/Immutable/Dictionary/index.js"));
var import_core = require("./core.js");
var import_excl_forEach = require("./excl-forEach.js");
var import_map = require("./map.js");
function bindAll(r, __trace) {
  return (s) => bindAll_(s, r, __trace);
}
function bindAll_(s, r, __trace) {
  return (0, import_core.chain_)(s, (k) => (0, import_map.map_)((0, import_excl_forEach.forEach_)(R.collect_(r(k), (k2, v) => [k2, v]), ([_, e]) => (0, import_map.map_)(e, (a) => [_, a])), (values) => {
    const res = {};
    for (const [k2, v] of values) {
      res[k2] = v;
    }
    return Object.assign(res, k);
  }), __trace);
}
function bindAllPar(r, __trace) {
  return (s) => bindAllPar_(s, r, __trace);
}
function bindAllPar_(s, r, __trace) {
  return (0, import_core.chain_)(s, (k) => (0, import_map.map_)((0, import_excl_forEach.forEachPar_)(R.collect_(r(k), (k2, v) => [k2, v]), ([_, e]) => (0, import_map.map_)(e, (a) => [_, a])), (values) => {
    const res = {};
    for (const [k2, v] of values) {
      res[k2] = v;
    }
    return Object.assign(res, k);
  }), __trace);
}
function bindAllParN(n, r, __trace) {
  return (s) => bindAllParN_(s, n, r, __trace);
}
function bindAllParN_(s, n, r, __trace) {
  return (0, import_core.chain_)(s, (k) => (0, import_map.map_)((0, import_excl_forEach.forEachParN_)(R.collect_(r(k), (k2, v) => [k2, v]), n, ([_, e]) => (0, import_map.map_)(e, (a) => [_, a])), (values) => {
    const res = {};
    for (const [k2, v] of values) {
      res[k2] = v;
    }
    return Object.assign(res, k);
  }), __trace);
}
module.exports = __toCommonJS(bindAll_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  bindAll,
  bindAllPar,
  bindAllParN,
  bindAllParN_,
  bindAllPar_,
  bindAll_
});
