// @effect-ts/system/Effect/collect.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var collect_exports = {};
__export(collect_exports, {
  collect: () => collect,
  collectPar: () => collectPar,
  collectParN: () => collectParN,
  collectParN_: () => collectParN_,
  collectPar_: () => collectPar_,
  collect_: () => collect_
});
var import_compact = require("../Collections/Immutable/Chunk/api/compact.js");
var import_excl_forEach = require("./excl-forEach.js");
var import_map = require("./map.js");
var import_optional = require("./optional.js");
function collect(f, __trace) {
  return (self) => collect_(self, f, __trace);
}
function collect_(self, f, __trace) {
  return (0, import_map.map_)((0, import_excl_forEach.forEach_)(self, (a) => (0, import_optional.optional)(f(a)), __trace), import_compact.compact);
}
function collectPar(f, __trace) {
  return (self) => collectPar_(self, f, __trace);
}
function collectPar_(self, f, __trace) {
  return (0, import_map.map_)((0, import_excl_forEach.forEachPar_)(self, (a) => (0, import_optional.optional)(f(a)), __trace), import_compact.compact);
}
function collectParN_(self, n, f, __trace) {
  return (0, import_map.map_)((0, import_excl_forEach.forEachParN_)(self, n, (a) => (0, import_optional.optional)(f(a)), __trace), import_compact.compact);
}
function collectParN(n, f, __trace) {
  return (self) => collectParN_(self, n, f, __trace);
}
module.exports = __toCommonJS(collect_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  collect,
  collectPar,
  collectParN,
  collectParN_,
  collectPar_,
  collect_
});
