// @effect-ts/system/Effect/excl-forEach-promise.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var excl_forEach_promise_exports = {};
__export(excl_forEach_promise_exports, {
  Promise: () => import_promise.Promise,
  await: () => import_await.await,
  fail: () => import_fail.fail,
  halt: () => import_halt.halt,
  interruptAs: () => import_interruptAs.interruptAs,
  make: () => import_make.make,
  succeed: () => import_succeed.succeed,
  unsafeDone: () => import_unsafeDone.unsafeDone,
  unsafeMake: () => import_unsafeMake.unsafeMake
});
var import_make = require("../Promise/make.js");
var import_fail = require("../Promise/fail.js");
var import_succeed = require("../Promise/succeed.js");
var import_await = require("../Promise/await.js");
var import_promise = require("../Promise/promise.js");
var import_halt = require("../Promise/halt.js");
var import_unsafeDone = require("../Promise/unsafeDone.js");
var import_unsafeMake = require("../Promise/unsafeMake.js");
var import_interruptAs = require("../Promise/interruptAs.js");
module.exports = __toCommonJS(excl_forEach_promise_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Promise,
  await,
  fail,
  halt,
  interruptAs,
  make,
  succeed,
  unsafeDone,
  unsafeMake
});
