// @effect-ts/system/Effect/ExecutionStrategy.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var ExecutionStrategy_exports = {};
__export(ExecutionStrategy_exports, {
  Parallel: () => Parallel,
  ParallelN: () => ParallelN,
  Sequential: () => Sequential,
  parallel: () => parallel,
  parallelN: () => parallelN,
  sequential: () => sequential
});
class Sequential {
  constructor() {
    this._tag = "Sequential";
  }
}
class Parallel {
  constructor() {
    this._tag = "Parallel";
  }
}
class ParallelN {
  constructor(n) {
    this.n = n;
    this._tag = "ParallelN";
  }
}
const sequential = /* @__PURE__ */ new Sequential();
const parallel = /* @__PURE__ */ new Parallel();
function parallelN(n) {
  return new ParallelN(n);
}
module.exports = __toCommonJS(ExecutionStrategy_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Parallel,
  ParallelN,
  Sequential,
  parallel,
  parallelN,
  sequential
});
