// @effect-ts/system/Effect/filterOr.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var filterOr_exports = {};
__export(filterOr_exports, {
  filterOrDie: () => filterOrDie,
  filterOrDieMessage: () => filterOrDieMessage,
  filterOrDieMessage_: () => filterOrDieMessage_,
  filterOrDie_: () => filterOrDie_,
  filterOrElse: () => filterOrElse,
  filterOrElse_: () => filterOrElse_,
  filterOrFail: () => filterOrFail,
  filterOrFail_: () => filterOrFail_
});
var import_Cause = require("../Cause/index.js");
var import_Function = require("../Function/index.js");
var import_core = require("./core.js");
var import_die = require("./die.js");
var import_fail = require("./fail.js");
function filterOrDie(p, dieWith, __trace) {
  return (fa) => filterOrDie_(fa, p, dieWith, __trace);
}
function filterOrDie_(fa, p, dieWith, __trace) {
  return filterOrElse_(fa, p, (x) => (0, import_die.die)(dieWith(x)), __trace);
}
function filterOrFail(p, failWith, __trace) {
  return (fa) => filterOrFail_(fa, p, failWith, __trace);
}
function filterOrFail_(fa, p, failWith, __trace) {
  return filterOrElse_(fa, p, (x) => (0, import_fail.fail)(failWith(x)), __trace);
}
function filterOrElse(p, or, __trace) {
  return (fa) => filterOrElse_(fa, p, or, __trace);
}
function filterOrElse_(fa, p, or, __trace) {
  return (0, import_core.chain_)(fa, (a) => p(a) ? (0, import_core.succeed)(a, __trace) : (0, import_core.suspend)(() => or(a), __trace));
}
function filterOrDieMessage(p, message, __trace) {
  return (fa) => filterOrDieMessage_(fa, p, message, __trace);
}
function filterOrDieMessage_(fa, p, message, __trace) {
  return filterOrDie_(fa, p, (a) => new import_Cause.RuntimeError(message(a)), __trace);
}
module.exports = __toCommonJS(filterOr_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  filterOrDie,
  filterOrDieMessage,
  filterOrDieMessage_,
  filterOrDie_,
  filterOrElse,
  filterOrElse_,
  filterOrFail,
  filterOrFail_
});
