// @effect-ts/system/Effect/has.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var has_exports = {};
__export(has_exports, {
  accessService: () => accessService,
  accessServiceM: () => accessServiceM,
  accessServices: () => accessServices,
  accessServicesM: () => accessServicesM,
  accessServicesT: () => accessServicesT,
  accessServicesTM: () => accessServicesTM,
  provideService: () => provideService,
  provideServiceM: () => provideServiceM,
  provideServiceM_: () => provideServiceM_,
  provideService_: () => provideService_,
  replaceService: () => replaceService,
  replaceServiceM: () => replaceServiceM,
  replaceServiceM_: () => replaceServiceM_,
  replaceService_: () => replaceService_,
  service: () => service,
  services: () => services
});
var A = __toESM(require("../Collections/Immutable/Array/index.js"));
var R = __toESM(require("../Collections/Immutable/Dictionary/index.js"));
var core = __toESM(require("../Effect/core.js"));
var import_Has = require("../Has/index.js");
var import_Tracing = require("../Tracing/index.js");
function accessServicesM(s) {
  return (f, __trace) => core.accessM((r) => f(R.map_(s, (v) => r[v.key])), __trace);
}
function accessServicesTM(...s) {
  return (f, __trace) => core.accessM((r) => f(...A.map_(s, (v) => r[v.key])), __trace);
}
function accessServicesT(...s) {
  return (f, __trace) => core.access((r) => f(...A.map_(s, (v) => r[v.key])), __trace);
}
function accessServices(s) {
  return (f, __trace) => core.access((r) => f(R.map_(s, (v) => r[v.key])), __trace);
}
function accessServiceM(s) {
  return (f, __trace) => core.accessM((r) => f(r[s.key]), __trace);
}
function accessService(s) {
  return (f, __trace) => accessServiceM(s)((a) => core.succeed(f(a)), __trace);
}
function service(s, __trace) {
  return accessServiceM(s)(core.succeed, __trace);
}
function services(...s) {
  return core.access((r) => s.map((tag) => tag.read(r)), (0, import_Tracing.accessCallTrace)());
}
function provideServiceM(_) {
  return (service2, __trace) => (ma) => core.accessM((r) => core.chain_(service2, (t) => core.provideAll_(ma, (0, import_Has.mergeEnvironments)(_, r, t), __trace)));
}
function provideServiceM_(ma, _, service2, __trace) {
  return core.accessM((r) => core.chain_(service2, (t) => core.provideAll_(ma, (0, import_Has.mergeEnvironments)(_, r, t), __trace)));
}
function provideService(_) {
  return (service2, __trace) => (ma) => provideServiceM(_)(core.succeed(service2), __trace)(ma);
}
function provideService_(ma, _, service2, __trace) {
  return provideServiceM(_)(core.succeed(service2), __trace)(ma);
}
function replaceServiceM(_, f, __trace) {
  return (ma) => accessServiceM(_)((t) => provideServiceM(_)(f(t), __trace)(ma));
}
function replaceServiceM_(ma, _, f, __trace) {
  return accessServiceM(_)((t) => provideServiceM(_)(f(t), __trace)(ma));
}
function replaceService(_, f, __trace) {
  return (ma) => replaceService_(ma, _, f, __trace);
}
function replaceService_(ma, _, f, __trace) {
  return accessServiceM(_)((t) => provideServiceM(_)(core.succeedWith(() => f(t)), __trace)(ma));
}
module.exports = __toCommonJS(has_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  accessService,
  accessServiceM,
  accessServices,
  accessServicesM,
  accessServicesT,
  accessServicesTM,
  provideService,
  provideServiceM,
  provideServiceM_,
  provideService_,
  replaceService,
  replaceServiceM,
  replaceServiceM_,
  replaceService_,
  service,
  services
});
