// @effect-ts/system/Effect/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var Effect_exports = {};
__export(Effect_exports, {
  collectAll: () => import_excl_forEach.collectAll,
  collectAllPar: () => import_excl_forEach.collectAllPar,
  collectAllParN: () => import_excl_forEach.collectAllParN,
  collectAllParN_: () => import_excl_forEach.collectAllParN_,
  collectAllSuccesses: () => import_excl_forEach.collectAllSuccesses,
  collectAllSuccessesPar: () => import_excl_forEach.collectAllSuccessesPar,
  collectAllSuccessesParN: () => import_excl_forEach.collectAllSuccessesParN,
  collectAllUnit: () => import_excl_forEach.collectAllUnit,
  collectAllUnitPar: () => import_excl_forEach.collectAllUnitPar,
  collectAllUnitParN: () => import_excl_forEach.collectAllUnitParN,
  collectAllUnitParN_: () => import_excl_forEach.collectAllUnitParN_,
  collectAllWith: () => import_excl_forEach.collectAllWith,
  collectAllWithPar: () => import_excl_forEach.collectAllWithPar,
  collectAllWithParN: () => import_excl_forEach.collectAllWithParN,
  collectAllWithParN_: () => import_excl_forEach.collectAllWithParN_,
  collectAllWithPar_: () => import_excl_forEach.collectAllWithPar_,
  collectAllWith_: () => import_excl_forEach.collectAllWith_,
  forEach: () => import_excl_forEach.forEach,
  forEachExec: () => import_excl_forEach.forEachExec,
  forEachExec_: () => import_excl_forEach.forEachExec_,
  forEachPar: () => import_excl_forEach.forEachPar,
  forEachParN: () => import_excl_forEach.forEachParN,
  forEachParN_: () => import_excl_forEach.forEachParN_,
  forEachParWithIndex: () => import_excl_forEach.forEachParWithIndex,
  forEachParWithIndexN: () => import_excl_forEach.forEachParWithIndexN,
  forEachParWithIndexN_: () => import_excl_forEach.forEachParWithIndexN_,
  forEachParWithIndex_: () => import_excl_forEach.forEachParWithIndex_,
  forEachPar_: () => import_excl_forEach.forEachPar_,
  forEachUnit: () => import_excl_forEach.forEachUnit,
  forEachUnitPar: () => import_excl_forEach.forEachUnitPar,
  forEachUnitParN: () => import_excl_forEach.forEachUnitParN,
  forEachUnitParN_: () => import_excl_forEach.forEachUnitParN_,
  forEachUnitPar_: () => import_excl_forEach.forEachUnitPar_,
  forEachUnit_: () => import_excl_forEach.forEachUnit_,
  forEachWithIndex: () => import_excl_forEach.forEachWithIndex,
  forEachWithIndex_: () => import_excl_forEach.forEachWithIndex_,
  forEach_: () => import_excl_forEach.forEach_,
  forkManaged: () => import_excl_forEach.forkManaged
});
var import_Operator = require("../Operator/index.js");
__reExport(Effect_exports, require("./absolve.js"));
__reExport(Effect_exports, require("./absorbWith.js"));
__reExport(Effect_exports, require("./andThen.js"));
__reExport(Effect_exports, require("./ap.js"));
__reExport(Effect_exports, require("./as.js"));
__reExport(Effect_exports, require("./asService.js"));
__reExport(Effect_exports, require("./asSome.js"));
__reExport(Effect_exports, require("./asSomeError.js"));
__reExport(Effect_exports, require("./asUnit.js"));
__reExport(Effect_exports, require("./awaitAllChildren.js"));
__reExport(Effect_exports, require("./bimap.js"));
__reExport(Effect_exports, require("./bindAll.js"));
__reExport(Effect_exports, require("./bracket.js"));
__reExport(Effect_exports, require("./bracketExit.js"));
__reExport(Effect_exports, require("./bracketFiber.js"));
__reExport(Effect_exports, require("./bracketOnError.js"));
__reExport(Effect_exports, require("./cached.js"));
__reExport(Effect_exports, require("./cachedInvalidate.js"));
__reExport(Effect_exports, require("./Canceler.js"));
__reExport(Effect_exports, require("./catchAll.js"));
__reExport(Effect_exports, require("./catchAllCause.js"));
__reExport(Effect_exports, require("./catchAllDefect.js"));
__reExport(Effect_exports, require("./catchSome.js"));
__reExport(Effect_exports, require("./catchSomeCause_.js"));
__reExport(Effect_exports, require("./catchSomeDefect.js"));
__reExport(Effect_exports, require("./cause.js"));
__reExport(Effect_exports, require("./Cb.js"));
__reExport(Effect_exports, require("./chainError.js"));
__reExport(Effect_exports, require("./collect.js"));
__reExport(Effect_exports, require("./commons.js"));
__reExport(Effect_exports, require("./compose.js"));
__reExport(Effect_exports, require("./cond.js"));
__reExport(Effect_exports, require("./continueOrFail.js"));
__reExport(Effect_exports, require("./core-scope.js"));
__reExport(Effect_exports, require("./core.js"));
__reExport(Effect_exports, require("./delay.js"));
__reExport(Effect_exports, require("./derive.js"));
__reExport(Effect_exports, require("./descriptor.js"));
__reExport(Effect_exports, require("./die.js"));
__reExport(Effect_exports, require("./dieMessage.js"));
__reExport(Effect_exports, require("./do.js"));
__reExport(Effect_exports, require("./done.js"));
__reExport(Effect_exports, require("./dropWhile.js"));
__reExport(Effect_exports, require("./effect.js"));
__reExport(Effect_exports, require("./effectAsync.js"));
__reExport(Effect_exports, require("./effectAsyncInterrupt.js"));
__reExport(Effect_exports, require("./effectAsyncM.js"));
__reExport(Effect_exports, require("./effectMaybeAsyncInterrupt.js"));
__reExport(Effect_exports, require("./either.js"));
__reExport(Effect_exports, require("./ensuring.js"));
__reExport(Effect_exports, require("./ensuringChildren.js"));
__reExport(Effect_exports, require("./environment.js"));
__reExport(Effect_exports, require("./eventually.js"));
__reExport(Effect_exports, require("./ExecutionStrategy.js"));
__reExport(Effect_exports, require("./fail.js"));
__reExport(Effect_exports, require("./fiberId.js"));
__reExport(Effect_exports, require("./filter.js"));
__reExport(Effect_exports, require("./filterOr.js"));
__reExport(Effect_exports, require("./first.js"));
__reExport(Effect_exports, require("./firstSuccessOf.js"));
__reExport(Effect_exports, require("./flatten.js"));
__reExport(Effect_exports, require("./flattenErrorOption.js"));
__reExport(Effect_exports, require("./flip.js"));
__reExport(Effect_exports, require("./flipWith.js"));
__reExport(Effect_exports, require("./fold.js"));
__reExport(Effect_exports, require("./foldCause.js"));
__reExport(Effect_exports, require("./foldM.js"));
__reExport(Effect_exports, require("./forever.js"));
__reExport(Effect_exports, require("./forkAll.js"));
__reExport(Effect_exports, require("./forkAs.js"));
__reExport(Effect_exports, require("./forkManaged.js"));
__reExport(Effect_exports, require("./forkWithErrorHandler.js"));
__reExport(Effect_exports, require("./fromEither.js"));
__reExport(Effect_exports, require("./fromFiber.js"));
__reExport(Effect_exports, require("./fromNodeCb.js"));
__reExport(Effect_exports, require("./fromOption.js"));
__reExport(Effect_exports, require("./fromPredicate.js"));
__reExport(Effect_exports, require("./gen.js"));
__reExport(Effect_exports, require("./get.js"));
__reExport(Effect_exports, require("./getOrFail.js"));
__reExport(Effect_exports, require("./has.js"));
__reExport(Effect_exports, require("./head.js"));
__reExport(Effect_exports, require("./ifM.js"));
__reExport(Effect_exports, require("./ignore.js"));
__reExport(Effect_exports, require("./in.js"));
__reExport(Effect_exports, require("./interop.js"));
__reExport(Effect_exports, require("./interruptAllChildren.js"));
__reExport(Effect_exports, require("./interruption.js"));
__reExport(Effect_exports, require("./isFailure.js"));
__reExport(Effect_exports, require("./isSuccess.js"));
__reExport(Effect_exports, require("./iterate.js"));
__reExport(Effect_exports, require("./join.js"));
__reExport(Effect_exports, require("./leftOrFail.js"));
__reExport(Effect_exports, require("./loop.js"));
__reExport(Effect_exports, require("./map.js"));
__reExport(Effect_exports, require("./mapBoth.js"));
__reExport(Effect_exports, require("./mapError.js"));
__reExport(Effect_exports, require("./mapErrorCause.js"));
__reExport(Effect_exports, require("./mapN.js"));
__reExport(Effect_exports, require("./mapTryCatch.js"));
__reExport(Effect_exports, require("./memoize.js"));
__reExport(Effect_exports, require("./merge.js"));
__reExport(Effect_exports, require("./mergeAll.js"));
__reExport(Effect_exports, require("./never.js"));
__reExport(Effect_exports, require("./none.js"));
__reExport(Effect_exports, require("./once.js"));
__reExport(Effect_exports, require("./onExit.js"));
__reExport(Effect_exports, require("./onFirst.js"));
__reExport(Effect_exports, require("./onLeft.js"));
__reExport(Effect_exports, require("./onRight.js"));
__reExport(Effect_exports, require("./onSecond.js"));
__reExport(Effect_exports, require("./onTermination.js"));
__reExport(Effect_exports, require("./option.js"));
__reExport(Effect_exports, require("./optional.js"));
__reExport(Effect_exports, require("./orDie.js"));
__reExport(Effect_exports, require("./orDieKeep.js"));
__reExport(Effect_exports, require("./orDieWith.js"));
__reExport(Effect_exports, require("./orElse.js"));
__reExport(Effect_exports, require("./orElseEither.js"));
__reExport(Effect_exports, require("./orElseFail.js"));
__reExport(Effect_exports, require("./orElseOptional.js"));
__reExport(Effect_exports, require("./orElseSucceed.js"));
__reExport(Effect_exports, require("./parallelErrors.js"));
__reExport(Effect_exports, require("./partition.js"));
__reExport(Effect_exports, require("./primitives.js"));
__reExport(Effect_exports, require("./promise.js"));
__reExport(Effect_exports, require("./provide.js"));
__reExport(Effect_exports, require("./provideManaged.js"));
__reExport(Effect_exports, require("./provideSome.js"));
__reExport(Effect_exports, require("./provideSomeLayer.js"));
__reExport(Effect_exports, require("./race.js"));
__reExport(Effect_exports, require("./raceAll.js"));
__reExport(Effect_exports, require("./reduce.js"));
__reExport(Effect_exports, require("./reduceAll.js"));
__reExport(Effect_exports, require("./refailWithTrace.js"));
__reExport(Effect_exports, require("./refineOrDie.js"));
__reExport(Effect_exports, require("./reject.js"));
__reExport(Effect_exports, require("./repeat.js"));
__reExport(Effect_exports, require("./repeatN.js"));
__reExport(Effect_exports, require("./repeatUntil.js"));
__reExport(Effect_exports, require("./repeatWhile.js"));
__reExport(Effect_exports, require("./replicate.js"));
__reExport(Effect_exports, require("./require.js"));
__reExport(Effect_exports, require("./reserve.js"));
__reExport(Effect_exports, require("./restrictions.js"));
__reExport(Effect_exports, require("./resurrect.js"));
__reExport(Effect_exports, require("./retry.js"));
__reExport(Effect_exports, require("./retryUntil.js"));
__reExport(Effect_exports, require("./retryWhile.js"));
__reExport(Effect_exports, require("./right.js"));
__reExport(Effect_exports, require("./runtime.js"));
__reExport(Effect_exports, require("./sandbox.js"));
__reExport(Effect_exports, require("./sandboxWith.js"));
__reExport(Effect_exports, require("./scopeWith.js"));
__reExport(Effect_exports, require("./second.js"));
__reExport(Effect_exports, require("./sleep.js"));
__reExport(Effect_exports, require("./some.js"));
__reExport(Effect_exports, require("./someOrElse.js"));
__reExport(Effect_exports, require("./someOrElseM.js"));
__reExport(Effect_exports, require("./someOrFail.js"));
__reExport(Effect_exports, require("./someOrFailException.js"));
__reExport(Effect_exports, require("./struct.js"));
__reExport(Effect_exports, require("./summarized.js"));
__reExport(Effect_exports, require("./swap.js"));
__reExport(Effect_exports, require("./tap.js"));
__reExport(Effect_exports, require("./tapBoth.js"));
__reExport(Effect_exports, require("./tapCause.js"));
__reExport(Effect_exports, require("./tapEither.js"));
__reExport(Effect_exports, require("./tapError.js"));
__reExport(Effect_exports, require("./tapExit.js"));
__reExport(Effect_exports, require("./timed.js"));
__reExport(Effect_exports, require("./timeout.js"));
__reExport(Effect_exports, require("./timeoutFail.js"));
__reExport(Effect_exports, require("./timeoutFailCause.js"));
__reExport(Effect_exports, require("./timeoutTo.js"));
__reExport(Effect_exports, require("./to.js"));
__reExport(Effect_exports, require("./toLayer.js"));
__reExport(Effect_exports, require("./toLeft.js"));
__reExport(Effect_exports, require("./toManaged.js"));
__reExport(Effect_exports, require("./traceMask.js"));
__reExport(Effect_exports, require("./tryCatchOption.js"));
__reExport(Effect_exports, require("./tuple.js"));
__reExport(Effect_exports, require("./uncause.js"));
__reExport(Effect_exports, require("./union.js"));
__reExport(Effect_exports, require("./unless.js"));
__reExport(Effect_exports, require("./unoption.js"));
__reExport(Effect_exports, require("./unrefine.js"));
__reExport(Effect_exports, require("./unsandbox.js"));
__reExport(Effect_exports, require("./unsome.js"));
__reExport(Effect_exports, require("./updateService.js"));
__reExport(Effect_exports, require("./validate.js"));
__reExport(Effect_exports, require("./validateFirst.js"));
__reExport(Effect_exports, require("./when.js"));
__reExport(Effect_exports, require("./whenCase.js"));
__reExport(Effect_exports, require("./whenM.js"));
__reExport(Effect_exports, require("./withChildren.js"));
__reExport(Effect_exports, require("./zip.js"));
__reExport(Effect_exports, require("./zipPar.js"));
__reExport(Effect_exports, require("./zips.js"));
__reExport(Effect_exports, require("./zipWith.js"));
__reExport(Effect_exports, require("./zipWithPar.js"));
var import_excl_forEach = require("./excl-forEach.js");
module.exports = __toCommonJS(Effect_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  collectAll,
  collectAllPar,
  collectAllParN,
  collectAllParN_,
  collectAllSuccesses,
  collectAllSuccessesPar,
  collectAllSuccessesParN,
  collectAllUnit,
  collectAllUnitPar,
  collectAllUnitParN,
  collectAllUnitParN_,
  collectAllWith,
  collectAllWithPar,
  collectAllWithParN,
  collectAllWithParN_,
  collectAllWithPar_,
  collectAllWith_,
  forEach,
  forEachExec,
  forEachExec_,
  forEachPar,
  forEachParN,
  forEachParN_,
  forEachParWithIndex,
  forEachParWithIndexN,
  forEachParWithIndexN_,
  forEachParWithIndex_,
  forEachPar_,
  forEachUnit,
  forEachUnitPar,
  forEachUnitParN,
  forEachUnitParN_,
  forEachUnitPar_,
  forEachUnit_,
  forEachWithIndex,
  forEachWithIndex_,
  forEach_,
  forkManaged
});
