// @effect-ts/system/Effect/repeat.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var repeat_exports = {};
__export(repeat_exports, {
  repeat: () => repeat,
  repeatOrElse: () => repeatOrElse,
  repeatOrElseEither: () => repeatOrElseEither,
  repeatOrElseEither_: () => repeatOrElseEither_,
  repeatOrElse_: () => repeatOrElse_,
  repeat_: () => repeat_
});
var E = __toESM(require("../Either/index.js"));
var import_Function = require("../Function/index.js");
var O = __toESM(require("../Option/index.js"));
var S = __toESM(require("../Schedule/index.js"));
var import_core = require("./core.js");
var import_fail = require("./fail.js");
var import_foldM = require("./foldM.js");
var map = __toESM(require("./map.js"));
var import_orDie = require("./orDie.js");
function repeatOrElseEither_(self, schedule, orElse, __trace) {
  return (0, import_core.chain)((driver) => {
    function loop(a) {
      return (0, import_foldM.foldM)(() => map.map_((0, import_orDie.orDie)(driver.last), E.right), (b) => (0, import_foldM.foldM)((e) => map.map_(orElse(e, O.some(b)), E.left), (a2) => loop(a2))(self))(driver.next(a));
    }
    return (0, import_foldM.foldM)((e) => map.map_(orElse(e, O.none), E.left), (a) => loop(a), __trace)(self);
  })(S.driver(schedule));
}
function repeatOrElseEither(schedule, orElse, __trace) {
  return (self) => repeatOrElseEither_(self, schedule, orElse, __trace);
}
function repeatOrElse_(self, schedule, orElse, __trace) {
  return map.map_(repeatOrElseEither_(self, schedule, orElse, __trace), E.merge);
}
function repeatOrElse(schedule, orElse, __trace) {
  return (self) => repeatOrElse_(self, schedule, orElse, __trace);
}
function repeat_(self, schedule, __trace) {
  return repeatOrElse_(self, schedule, (e) => (0, import_fail.fail)(e), __trace);
}
function repeat(schedule, __trace) {
  return (self) => repeat_(self, schedule, __trace);
}
module.exports = __toCommonJS(repeat_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  repeat,
  repeatOrElse,
  repeatOrElseEither,
  repeatOrElseEither_,
  repeatOrElse_,
  repeat_
});
