// @effect-ts/system/Effect/validateFirst.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var validateFirst_exports = {};
__export(validateFirst_exports, {
  validateFirst: () => validateFirst,
  validateFirstPar: () => validateFirstPar,
  validateFirstParN: () => validateFirstParN,
  validateFirstParN_: () => validateFirstParN_,
  validateFirstPar_: () => validateFirstPar_,
  validateFirst_: () => validateFirst_
});
var import_excl_forEach = require("./excl-forEach.js");
var import_flip = require("./flip.js");
function validateFirst_(i, f, __trace) {
  return (0, import_flip.flip)((0, import_excl_forEach.forEach_)(i, (a) => (0, import_flip.flip)(f(a)), __trace));
}
function validateFirst(f, __trace) {
  return (i) => validateFirst_(i, f, __trace);
}
function validateFirstPar_(i, f, __trace) {
  return (0, import_flip.flip)((0, import_excl_forEach.forEachPar_)(i, (a) => (0, import_flip.flip)(f(a)), __trace));
}
function validateFirstPar(f, __trace) {
  return (i) => validateFirstPar_(i, f, __trace);
}
function validateFirstParN_(i, n, f, __trace) {
  return (0, import_flip.flip)((0, import_excl_forEach.forEachParN_)(i, n, (a) => (0, import_flip.flip)(f(a)), __trace));
}
function validateFirstParN(n, f, __trace) {
  return (i) => validateFirstParN_(i, n, f, __trace);
}
module.exports = __toCommonJS(validateFirst_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  validateFirst,
  validateFirstPar,
  validateFirstParN,
  validateFirstParN_,
  validateFirstPar_,
  validateFirst_
});
