// @effect-ts/core/_esm/index.js
// ets_tracing: off
import "./Operator/index.js";
import * as Array_1 from "./Collections/Immutable/Array/index.js";
export { Array_1 as Array };
import * as Associative_1 from "./Associative/index.js";
export { Associative_1 as Associative };
import * as Async_1 from "./Async/index.js";
export { Async_1 as Async };
import * as Boolean_1 from "./Boolean/index.js";
export { Boolean_1 as Boolean };
import * as Branded_1 from "./Branded/index.js";
export { Branded_1 as Branded };
import * as Chunk_1 from "./Collections/Immutable/Chunk/index.js";
export { Chunk_1 as Chunk };
import * as Closure_1 from "./Closure/index.js";
export { Closure_1 as Closure };
import * as Commutative_1 from "./Commutative/index.js";
export { Commutative_1 as Commutative };
import * as Const_1 from "./Const/index.js";
export { Const_1 as Const };
import * as Effect_1 from "./Effect/index.js";
export { Effect_1 as Effect };
import * as Cause_1 from "./Effect/Cause/index.js";
export { Cause_1 as Cause };
import * as PrettyCause_1 from "./Effect/Cause/Pretty/index.js";
export { PrettyCause_1 as PrettyCause };
import * as Clock_1 from "./Effect/Clock/index.js";
export { Clock_1 as Clock };
import * as Exit_1 from "./Effect/Exit/index.js";
export { Exit_1 as Exit };
import * as Fiber_1 from "./Effect/Fiber/index.js";
export { Fiber_1 as Fiber };
import * as FiberRef_1 from "./Effect/FiberRef/index.js";
export { FiberRef_1 as FiberRef };
import * as Managed_1 from "./Effect/Managed/index.js";
export { Managed_1 as Managed };
import * as Promise_1 from "./Effect/Promise/index.js";
export { Promise_1 as Promise };
import * as Queue_1 from "./Effect/Queue/index.js";
export { Queue_1 as Queue };
import * as Random_1 from "./Effect/Random/index.js";
export { Random_1 as Random };
import * as Ref_1 from "./Effect/Ref/index.js";
export { Ref_1 as Ref };
import * as RefM_1 from "./Effect/RefM/index.js";
export { RefM_1 as RefM };
import * as Schedule_1 from "./Effect/Schedule/index.js";
export { Schedule_1 as Schedule };
import * as Scope_1 from "./Effect/Scope/index.js";
export { Scope_1 as Scope };
import * as Semaphore_1 from "./Effect/Semaphore/index.js";
export { Semaphore_1 as Semaphore };
import * as Stream_1 from "./Effect/Stream/index.js";
export { Stream_1 as Stream };
import * as BufferedPull_1 from "./Effect/Stream/BufferedPull/index.js";
export { BufferedPull_1 as BufferedPull };
import * as Pull_1 from "./Effect/Stream/Pull/index.js";
export { Pull_1 as Pull };
import * as Push_1 from "./Effect/Stream/Push/index.js";
export { Push_1 as Push };
import * as Sink_1 from "./Effect/Stream/Sink/index.js";
export { Sink_1 as Sink };
import * as Take_1 from "./Effect/Stream/Take/index.js";
export { Take_1 as Take };
import * as Transducer_1 from "./Effect/Stream/Transducer/index.js";
export { Transducer_1 as Transducer };
import * as Supervisor_1 from "./Effect/Supervisor/index.js";
export { Supervisor_1 as Supervisor };
import * as Either_1 from "./Either/index.js";
export { Either_1 as Either };
import * as EitherT_1 from "./EitherT/index.js";
export { EitherT_1 as EitherT };
import * as Equal_1 from "./Equal/index.js";
export { Equal_1 as Equal };
import * as FreeAssociative_1 from "./FreeAssociative/index.js";
export { FreeAssociative_1 as FreeAssociative };
import * as Function_1 from "./Function/index.js";
export { Function_1 as Function };
import * as Has_1 from "./Has/index.js";
export { Has_1 as Has };
import * as Hash_1 from "./Hash/index.js";
export { Hash_1 as Hash };
import * as HashMap_1 from "./Collections/Immutable/HashMap/index.js";
export { HashMap_1 as HashMap };
import * as HashSet_1 from "./Collections/Immutable/HashSet/index.js";
export { HashSet_1 as HashSet };
import * as IO_1 from "./IO/index.js";
export { IO_1 as IO };
import * as Id_1 from "./Id/index.js";
export { Id_1 as Id };
import * as Identity_1 from "./Identity/index.js";
export { Identity_1 as Identity };
import * as IndexedT_1 from "./IndexedT/index.js";
export { IndexedT_1 as IndexedT };
import * as InvariantT_1 from "./InvariantT/index.js";
export { InvariantT_1 as InvariantT };
import * as Inverse_1 from "./Inverse/index.js";
export { Inverse_1 as Inverse };
import * as Iterable_1 from "./Iterable/index.js";
export { Iterable_1 as Iterable };
import * as List_1 from "./Collections/Immutable/List/index.js";
export { List_1 as List };
import * as Map_1 from "./Collections/Immutable/Map/index.js";
export { Map_1 as Map };
import * as Newtype_1 from "./Newtype/index.js";
export { Newtype_1 as Newtype };
import * as NonEmptyArray_1 from "./Collections/Immutable/NonEmptyArray/index.js";
export { NonEmptyArray_1 as NonEmptyArray };
import * as Option_1 from "./Option/index.js";
export { Option_1 as Option };
import * as OptionT_1 from "./OptionT/index.js";
export { OptionT_1 as OptionT };
import * as Ord_1 from "./Ord/index.js";
export { Ord_1 as Ord };
import * as Ordering_1 from "./Ordering/index.js";
export { Ordering_1 as Ordering };
import * as Any_1 from "./Prelude/Any/index.js";
export { Any_1 as Any };
import * as Applicative_1 from "./Prelude/Applicative/index.js";
export { Applicative_1 as Applicative };
import * as Apply_1 from "./Prelude/Apply/index.js";
export { Apply_1 as Apply };
import * as AssociativeBoth_1 from "./Prelude/AssociativeBoth/index.js";
export { AssociativeBoth_1 as AssociativeBoth };
import * as AssociativeCompose_1 from "./Prelude/AssociativeCompose/index.js";
export { AssociativeCompose_1 as AssociativeCompose };
import * as AssociativeEither_1 from "./Prelude/AssociativeEither/index.js";
export { AssociativeEither_1 as AssociativeEither };
import * as AssociativeFlatten_1 from "./Prelude/AssociativeFlatten/index.js";
export { AssociativeFlatten_1 as AssociativeFlatten };
import * as Category_1 from "./Prelude/Category/index.js";
export { Category_1 as Category };
import * as ChainRec_1 from "./Prelude/ChainRec/index.js";
export { ChainRec_1 as ChainRec };
import * as CommutativeBoth_1 from "./Prelude/CommutativeBoth/index.js";
export { CommutativeBoth_1 as CommutativeBoth };
import * as CommutativeEither_1 from "./Prelude/CommutativeEither/index.js";
export { CommutativeEither_1 as CommutativeEither };
import * as Compact_1 from "./Prelude/Compact/index.js";
export { Compact_1 as Compact };
import * as Compactable_1 from "./Prelude/Compactable/index.js";
export { Compactable_1 as Compactable };
import * as Contravariant_1 from "./Prelude/Contravariant/index.js";
export { Contravariant_1 as Contravariant };
import * as Covariant_1 from "./Prelude/Covariant/index.js";
export { Covariant_1 as Covariant };
import * as CovariantWithIndex_1 from "./Prelude/CovariantWithIndex/index.js";
export { CovariantWithIndex_1 as CovariantWithIndex };
import * as DSL_1 from "./Prelude/DSL/index.js";
export { DSL_1 as DSL };
import * as Derive_1 from "./Prelude/Derive/index.js";
export { Derive_1 as Derive };
import * as Extend_1 from "./Prelude/Extend/index.js";
export { Extend_1 as Extend };
import * as Access_1 from "./Prelude/FX/Access/index.js";
export { Access_1 as Access };
import * as Fail_1 from "./Prelude/FX/Fail/index.js";
export { Fail_1 as Fail };
import * as Provide_1 from "./Prelude/FX/Provide/index.js";
export { Provide_1 as Provide };
import * as Run_1 from "./Prelude/FX/Run/index.js";
export { Run_1 as Run };
import * as Filter_1 from "./Prelude/Filter/index.js";
export { Filter_1 as Filter };
import * as FilterMap_1 from "./Prelude/FilterMap/index.js";
export { FilterMap_1 as FilterMap };
import * as FilterMapWithIndex_1 from "./Prelude/FilterMapWithIndex/index.js";
export { FilterMapWithIndex_1 as FilterMapWithIndex };
import * as FilterWithIndex_1 from "./Prelude/FilterWithIndex/index.js";
export { FilterWithIndex_1 as FilterWithIndex };
import * as Filterable_1 from "./Prelude/Filterable/index.js";
export { Filterable_1 as Filterable };
import * as FilterableWithIndex_1 from "./Prelude/FilterableWithIndex/index.js";
export { FilterableWithIndex_1 as FilterableWithIndex };
import * as FoldMap_1 from "./Prelude/FoldMap/index.js";
export { FoldMap_1 as FoldMap };
import * as FoldMapWithIndex_1 from "./Prelude/FoldMapWithIndex/index.js";
export { FoldMapWithIndex_1 as FoldMapWithIndex };
import * as Foldable_1 from "./Prelude/Foldable/index.js";
export { Foldable_1 as Foldable };
import * as FoldableWithIndex_1 from "./Prelude/FoldableWithIndex/index.js";
export { FoldableWithIndex_1 as FoldableWithIndex };
import * as ForEach_1 from "./Prelude/ForEach/index.js";
export { ForEach_1 as ForEach };
import * as ForEachWithIndex_1 from "./Prelude/ForEachWithIndex/index.js";
export { ForEachWithIndex_1 as ForEachWithIndex };
import * as HKT_1 from "./Prelude/HKT/index.js";
export { HKT_1 as HKT };
import * as IdentityBoth_1 from "./Prelude/IdentityBoth/index.js";
export { IdentityBoth_1 as IdentityBoth };
import * as IdentityEither_1 from "./Prelude/IdentityEither/index.js";
export { IdentityEither_1 as IdentityEither };
import * as IdentityFlatten_1 from "./Prelude/IdentityFlatten/index.js";
export { IdentityFlatten_1 as IdentityFlatten };
import * as Invariant_1 from "./Prelude/Invariant/index.js";
export { Invariant_1 as Invariant };
import * as Monad_1 from "./Prelude/Monad/index.js";
export { Monad_1 as Monad };
import * as None_1 from "./Prelude/None/index.js";
export { None_1 as None };
import * as Partition_1 from "./Prelude/Partition/index.js";
export { Partition_1 as Partition };
import * as PartitionMap_1 from "./Prelude/PartitionMap/index.js";
export { PartitionMap_1 as PartitionMap };
import * as PartitionMapWithIndex_1 from "./Prelude/PartitionMapWithIndex/index.js";
export { PartitionMapWithIndex_1 as PartitionMapWithIndex };
import * as PartitionWithIndex_1 from "./Prelude/PartitionWithIndex/index.js";
export { PartitionWithIndex_1 as PartitionWithIndex };
import * as Reduce_1 from "./Prelude/Reduce/index.js";
export { Reduce_1 as Reduce };
import * as ReduceRight_1 from "./Prelude/ReduceRight/index.js";
export { ReduceRight_1 as ReduceRight };
import * as ReduceRightWithIndex_1 from "./Prelude/ReduceRightWithIndex/index.js";
export { ReduceRightWithIndex_1 as ReduceRightWithIndex };
import * as ReduceWithIndex_1 from "./Prelude/ReduceWithIndex/index.js";
export { ReduceWithIndex_1 as ReduceWithIndex };
import * as Selective_1 from "./Prelude/Selective/index.js";
export { Selective_1 as Selective };
import * as Separate_1 from "./Prelude/Separate/index.js";
export { Separate_1 as Separate };
import * as Wiltable_1 from "./Prelude/Wiltable/index.js";
export { Wiltable_1 as Wiltable };
import * as WiltableWithIndex_1 from "./Prelude/WiltableWithIndex/index.js";
export { WiltableWithIndex_1 as WiltableWithIndex };
import * as Witherable_1 from "./Prelude/Witherable/index.js";
export { Witherable_1 as Witherable };
import * as WitherableWithIndex_1 from "./Prelude/WitherableWithIndex/index.js";
export { WitherableWithIndex_1 as WitherableWithIndex };
import * as Reader_1 from "./Reader/index.js";
export { Reader_1 as Reader };
import * as ReaderT_1 from "./ReaderT/index.js";
export { ReaderT_1 as ReaderT };
import * as Record_1 from "./Collections/Immutable/Dictionary/index.js";
export { Record_1 as Record };
import * as RedBlackTree_1 from "./Collections/Immutable/RedBlackTree/index.js";
export { RedBlackTree_1 as RedBlackTree };
import * as RoseTree_1 from "./RoseTree/index.js";
export { RoseTree_1 as RoseTree };
import * as Set_1 from "./Collections/Immutable/Set/index.js";
export { Set_1 as Set };
import * as Show_1 from "./Show/index.js";
export { Show_1 as Show };
import * as SortedSet_1 from "./Collections/Immutable/SortedSet/index.js";
export { SortedSet_1 as SortedSet };
import * as StateT_1 from "./StateT/index.js";
export { StateT_1 as StateT };
import * as ParametricStateT_1 from "./StateT/Parametric/index.js";
export { ParametricStateT_1 as ParametricStateT };
import * as String_1 from "./String/index.js";
export { String_1 as String };
import * as AtomicBoolean_1 from "./Support/AtomicBoolean/index.js";
export { AtomicBoolean_1 as AtomicBoolean };
import * as AtomicNumber_1 from "./Support/AtomicNumber/index.js";
export { AtomicNumber_1 as AtomicNumber };
import * as AtomicReference_1 from "./Support/AtomicReference/index.js";
export { AtomicReference_1 as AtomicReference };
import * as DoublyLinkedList_1 from "./Support/DoublyLinkedList/index.js";
export { DoublyLinkedList_1 as DoublyLinkedList };
import * as ImmutableQueue_1 from "./Support/ImmutableQueue/index.js";
export { ImmutableQueue_1 as ImmutableQueue };
import * as LinkedList_1 from "./Support/LinkedList/index.js";
export { LinkedList_1 as LinkedList };
import * as MutableQueue_1 from "./Support/MutableQueue/index.js";
export { MutableQueue_1 as MutableQueue };
import * as OneShot_1 from "./Support/OneShot/index.js";
export { OneShot_1 as OneShot };
import * as RingBuffer_1 from "./Support/RingBuffer/index.js";
export { RingBuffer_1 as RingBuffer };
import * as Sync_1 from "./Sync/index.js";
export { Sync_1 as Sync };
import * as SyncLayer_1 from "./Sync/Layer/index.js";
export { SyncLayer_1 as SyncLayer };
import * as Utils_1 from "./Utils/index.js";
export { Utils_1 as Utils };
import * as XPure_1 from "./XPure/index.js";
export { XPure_1 as XPure };
import * as XIO_1 from "./XPure/XIO/index.js";
export { XIO_1 as XIO };
import * as XReader_1 from "./XPure/XReader/index.js";
export { XReader_1 as XReader };
import * as XReaderT_1 from "./XPure/XReaderT/index.js";
export { XReaderT_1 as XReaderT };
import * as XState_1 from "./XPure/XState/index.js";
export { XState_1 as XState }; //

export { pipe } from "./Function/index.js";
