// @effect-ts/system/Managed/methods/mapN.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var mapN_exports = {};
__export(mapN_exports, {
  mapN: () => mapN,
  mapNPar: () => mapNPar,
  mapNParN: () => mapNParN,
  mapNParN_: () => mapNParN_,
  mapNPar_: () => mapNPar_,
  mapN_: () => mapN_
});
var import_core = require("../core.js");
var import_tuple = require("../tuple.js");
function mapN(f, __trace) {
  return (t) => mapN_(t, f, __trace);
}
function mapN_(t, f, __trace) {
  return (0, import_core.map_)((0, import_tuple.tuple)(...t.tuple), (x) => f(...x), __trace);
}
function mapNPar(f, __trace) {
  return (t) => mapNPar_(t, f, __trace);
}
function mapNPar_(t, f, __trace) {
  return (0, import_core.map_)((0, import_tuple.tuplePar)(...t.tuple), (x) => f(...x.tuple), __trace);
}
function mapNParN(n, f, __trace) {
  return (t) => mapNParN_(t, n, f, __trace);
}
function mapNParN_(t, n, f, __trace) {
  return (0, import_core.map_)((0, import_tuple.tupleParN)(n)(...t.tuple), (x) => f(...x.tuple), __trace);
}
module.exports = __toCommonJS(mapN_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  mapN,
  mapNPar,
  mapNParN,
  mapNParN_,
  mapNPar_,
  mapN_
});
