// @effect-ts/core/IndexedT/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var IndexedT_exports = {};
__export(IndexedT_exports, {
  Ix: () => Ix,
  indexedF: () => indexedF,
  makeIx: () => makeIx
});
var import_Operator = require("../Operator/index.js");
var import_Function = require("../Function/index.js");
var import_DSL = require("../Prelude/DSL/index.js");
var HKT = __toESM(require("../Prelude/HKT/index.js"));
class Ix {
  constructor(value) {
    this.value = value;
  }
}
function makeIx() {
  return (a) => new Ix(a);
}
function indexedF() {
  return (F) => indexed_(F);
}
function indexed_(F) {
  return {
    iof: () => (a) => makeIx()((0, import_DSL.succeedF)(F)(a)),
    ichain: (f) => (fa) => makeIx()((0, import_DSL.chainF)(F)((a) => f(a).value)(fa.value)),
    lift: () => (fa) => makeIx()(fa),
    lower: () => (fa) => fa.value,
    chain: (f) => (fa) => makeIx()((0, import_DSL.chainF)(F)((a) => f(a).value)(fa.value)),
    chainLower: (f) => (fa) => makeIx()((0, import_DSL.chainF)(F)(f)(fa.value))
  };
}
module.exports = __toCommonJS(IndexedT_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Ix,
  indexedF,
  makeIx
});
