// @effect-ts/system/Stream/Stream/forEach.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var forEach_exports = {};
__export(forEach_exports, {
  forEach: () => forEach,
  forEachChunk: () => forEachChunk,
  forEachChunkManaged: () => forEachChunkManaged,
  forEachChunkManaged_: () => forEachChunkManaged_,
  forEachChunk_: () => forEachChunk_,
  forEachManaged: () => forEachManaged,
  forEachManaged_: () => forEachManaged_,
  forEachWhile: () => forEachWhile,
  forEachWhileManaged: () => forEachWhileManaged,
  forEachWhileManaged_: () => forEachWhileManaged_,
  forEachWhile_: () => forEachWhile_,
  forEach_: () => forEach_
});
var SK = __toESM(require("../Sink/index.js"));
var import_run = require("./run.js");
var import_runManaged = require("./runManaged.js");
function forEach_(self, f) {
  return (0, import_run.run_)(self, SK.forEach(f));
}
function forEach(f) {
  return (self) => forEach_(self, f);
}
function forEachWhileManaged_(self, f) {
  return (0, import_runManaged.runManaged_)(self, SK.forEachWhile(f));
}
function forEachWhileManaged(f) {
  return (self) => forEachWhileManaged_(self, f);
}
function forEachChunk_(self, f) {
  return (0, import_run.run_)(self, SK.forEachChunk(f));
}
function forEachChunk(f) {
  return (self) => (0, import_run.run_)(self, SK.forEachChunk(f));
}
function forEachWhile_(self, f) {
  return (0, import_run.run_)(self, SK.forEachWhile(f));
}
function forEachWhile(f) {
  return (self) => forEachWhile_(self, f);
}
function forEachManaged_(self, f) {
  return (0, import_runManaged.runManaged_)(self, SK.forEach(f));
}
function forEachManaged(f) {
  return (self) => forEachManaged_(self, f);
}
function forEachChunkManaged_(self, f) {
  return (0, import_runManaged.runManaged_)(self, SK.forEachChunk(f));
}
function forEachChunkManaged(f) {
  return (self) => forEachChunkManaged_(self, f);
}
module.exports = __toCommonJS(forEach_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  forEach,
  forEachChunk,
  forEachChunkManaged,
  forEachChunkManaged_,
  forEachChunk_,
  forEachManaged,
  forEachManaged_,
  forEachWhile,
  forEachWhileManaged,
  forEachWhileManaged_,
  forEachWhile_,
  forEach_
});
