// @effect-ts/system/Stream/Stream/merge.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var merge_exports = {};
__export(merge_exports, {
  merge: () => merge,
  mergeEither: () => mergeEither,
  mergeEither_: () => mergeEither_,
  mergeTerminateEither: () => mergeTerminateEither,
  mergeTerminateEither_: () => mergeTerminateEither_,
  mergeTerminateLeft: () => mergeTerminateLeft,
  mergeTerminateLeft_: () => mergeTerminateLeft_,
  mergeTerminateRight: () => mergeTerminateRight,
  mergeTerminateRight_: () => mergeTerminateRight_,
  merge_: () => merge_
});
var E = __toESM(require("../../Either/index.js"));
var import_mergeWith = require("./mergeWith.js");
function merge_(self, that, strategy = "Both") {
  return (0, import_mergeWith.mergeWith_)(self, that, (a) => a, (b) => b, strategy);
}
function merge(that, strategy = "Both") {
  return (self) => merge_(self, that, strategy);
}
function mergeTerminateEither_(self, that) {
  return merge_(self, that, "Either");
}
function mergeTerminateEither(that) {
  return (self) => merge_(self, that, "Either");
}
function mergeTerminateLeft_(self, that) {
  return merge_(self, that, "Left");
}
function mergeTerminateLeft(that) {
  return (self) => merge_(self, that, "Left");
}
function mergeTerminateRight_(self, that) {
  return merge_(self, that, "Right");
}
function mergeTerminateRight(that) {
  return (self) => merge_(self, that, "Right");
}
function mergeEither_(self, that, strategy = "Both") {
  return (0, import_mergeWith.mergeWith_)(self, that, (l) => E.left(l), E.right, strategy);
}
function mergeEither(that, strategy = "Both") {
  return (self) => mergeEither_(self, that, strategy);
}
module.exports = __toCommonJS(merge_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  merge,
  mergeEither,
  mergeEither_,
  mergeTerminateEither,
  mergeTerminateEither_,
  mergeTerminateLeft,
  mergeTerminateLeft_,
  mergeTerminateRight,
  mergeTerminateRight_,
  merge_
});
