// @effect-ts/system/Managed/deps-core.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var deps_core_exports = {};
__export(deps_core_exports, {
  exitCollectAll: () => import_core.collectAll,
  exitCollectAllPar: () => import_core.collectAllPar,
  exitForeach: () => import_api.forEach,
  exitHalt: () => import_api.halt,
  exitInterrupt: () => import_api.interrupt,
  exitSucceed: () => import_core.succeed,
  exitUnit: () => import_core.unit,
  exitZipRight_: () => import_core.zipRight_
});
__reExport(deps_core_exports, require("../Effect/zips.js"));
__reExport(deps_core_exports, require("../Effect/bracketExit.js"));
__reExport(deps_core_exports, require("../Effect/core.js"));
__reExport(deps_core_exports, require("../Effect/core-scope.js"));
__reExport(deps_core_exports, require("../Effect/do.js"));
__reExport(deps_core_exports, require("../Effect/done.js"));
__reExport(deps_core_exports, require("../Effect/effect.js"));
__reExport(deps_core_exports, require("../Effect/commons.js"));
__reExport(deps_core_exports, require("../Effect/environment.js"));
__reExport(deps_core_exports, require("../Effect/ExecutionStrategy.js"));
__reExport(deps_core_exports, require("../Effect/fail.js"));
__reExport(deps_core_exports, require("../Effect/flatten.js"));
__reExport(deps_core_exports, require("../Effect/interruption.js"));
__reExport(deps_core_exports, require("../Effect/map.js"));
__reExport(deps_core_exports, require("../Effect/mapError.js"));
__reExport(deps_core_exports, require("../Effect/mapErrorCause.js"));
__reExport(deps_core_exports, require("../Effect/never.js"));
__reExport(deps_core_exports, require("../Effect/provideSome.js"));
__reExport(deps_core_exports, require("../Effect/sandbox.js"));
__reExport(deps_core_exports, require("../Effect/tap.js"));
__reExport(deps_core_exports, require("../Effect/zipWith.js"));
__reExport(deps_core_exports, require("../Effect/zipWithPar.js"));
__reExport(deps_core_exports, require("../Effect/zip.js"));
__reExport(deps_core_exports, require("../Effect/zips.js"));
var import_api = require("../Exit/api.js");
var import_core = require("../Exit/core.js");
module.exports = __toCommonJS(deps_core_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  exitCollectAll,
  exitCollectAllPar,
  exitForeach,
  exitHalt,
  exitInterrupt,
  exitSucceed,
  exitUnit,
  exitZipRight_
});
