// @effect-ts/system/Has/index.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var Has_exports = {};
__export(Has_exports, {
  mergeEnvironments: () => mergeEnvironments,
  replaceServiceIn: () => replaceServiceIn,
  replaceServiceIn_: () => replaceServiceIn_,
  service: () => service,
  tag: () => tag
});
var import_Operator = require("../Operator/index.js");
var import_Option = require("../Option/index.js");
function service(x) {
  return x;
}
const makeTag = (key = Symbol()) => ({
  _tag: "Tag",
  _T: void 0,
  key,
  has: (t) => ({
    [key]: t
  }),
  of: (t) => t,
  read: (r) => r[key],
  readOption: (r) => typeof r === "object" && r !== null ? (0, import_Option.fromNullable)(r[key]) : import_Option.none,
  refine: () => makeTag(key)
});
function tag(key) {
  return makeTag(key);
}
const replaceServiceIn = (_, f) => (r) => __spreadProps(__spreadValues({}, r), {
  [_.key]: f(r[_.key])
});
const replaceServiceIn_ = (r, _, f) => __spreadValues(__spreadValues({}, r), _.has(f(r[_.key])));
function mergeEnvironments(_, r, t) {
  return __spreadValues(__spreadValues({}, r), _.has(t));
}
module.exports = __toCommonJS(Has_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  mergeEnvironments,
  replaceServiceIn,
  replaceServiceIn_,
  service,
  tag
});
