// @effect-ts/system/RefM/effect.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var effect_exports = {};
__export(effect_exports, {
  as: () => import_as.as,
  asSomeError: () => import_asSomeError.asSomeError,
  as_: () => import_as.as_,
  bind: () => import_do.bind,
  bind_: () => import_do.bind_,
  chain: () => import_core.chain,
  chain_: () => import_core.chain_,
  do: () => import_do.do,
  fail: () => import_fail.fail,
  foldM: () => import_foldM.foldM,
  foldM_: () => import_foldM.foldM_,
  fromEither: () => import_fromEither.fromEither,
  ifM_: () => import_ifM.ifM_,
  map: () => import_map.map,
  mapError_: () => import_mapError.mapError_,
  map_: () => import_map.map_,
  succeed: () => import_core.succeed,
  toManaged: () => import_toManaged.toManaged
});
var import_as = require("../Effect/as.js");
var import_asSomeError = require("../Effect/asSomeError.js");
var import_core = require("../Effect/core.js");
var import_do = require("../Effect/do.js");
var import_fail = require("../Effect/fail.js");
var import_foldM = require("../Effect/foldM.js");
var import_fromEither = require("../Effect/fromEither.js");
var import_ifM = require("../Effect/ifM.js");
var import_map = require("../Effect/map.js");
var import_mapError = require("../Effect/mapError.js");
var import_toManaged = require("../Effect/toManaged.js");
module.exports = __toCommonJS(effect_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  as,
  asSomeError,
  as_,
  bind,
  bind_,
  chain,
  chain_,
  do: null,
  fail,
  foldM,
  foldM_,
  fromEither,
  ifM_,
  map,
  mapError_,
  map_,
  succeed,
  toManaged
});
