// @effect-ts/system/Ref/XRef.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var XRef_exports = {};
__export(XRef_exports, {
  Atomic: () => Atomic,
  Derived: () => Derived,
  DerivedAll: () => DerivedAll,
  TypeId: () => TypeId,
  concrete: () => concrete
});
var Tp = __toESM(require("../Collections/Immutable/Tuple/index.js"));
var E = __toESM(require("../Either/index.js"));
var import_Function = require("../Function/index.js");
var atomic = __toESM(require("./atomic.js"));
var T = __toESM(require("./effect.js"));
const TypeId = /* @__PURE__ */ Symbol();
class Atomic {
  constructor(value) {
    this.value = value;
    this._tag = "Atomic";
    this._typeId = TypeId;
    this.fold = this.fold.bind(this);
    this.foldAll = this.foldAll.bind(this);
    this.set = this.set.bind(this);
  }
  fold(_ea, _eb, ca, bd) {
    return new Derived((f) => f(this, (s) => bd(s), (c) => ca(c)));
  }
  foldAll(_ea, _eb, _ec, ca, bd) {
    return new DerivedAll((f) => f(this, (s) => bd(s), (c) => (s) => ca(c)(s)));
  }
  get get() {
    return T.succeedWith(() => this.value.get);
  }
  set(a) {
    return T.succeedWith(() => {
      this.value.set(a);
    });
  }
}
class Derived {
  constructor(use) {
    this.use = use;
    this._tag = "Derived";
    this._typeId = TypeId;
    this.fold = this.fold.bind(this);
    this.foldAll = this.foldAll.bind(this);
    this.set = this.set.bind(this);
  }
  fold(ea, eb, ca, bd) {
    return this.use((value, getEither, setEither) => new Derived((f) => f(value, (s) => E.fold_(getEither(s), (e) => E.left(eb(e)), bd), (c) => E.chain_(ca(c), (a) => E.fold_(setEither(a), (e) => E.left(ea(e)), E.right)))));
  }
  foldAll(ea, eb, ec, ca, _bd) {
    return this.use((value, getEither, setEither) => new DerivedAll((f) => f(value, (s) => E.fold_(getEither(s), (e) => E.left(eb(e)), E.right), (c) => (s) => E.chain_(E.fold_(getEither(s), (e) => E.left(ec(e)), ca(c)), (a) => E.fold_(setEither(a), (e) => E.left(ea(e)), E.right)))));
  }
  get get() {
    return this.use((value, getEither) => T.chain_(value.get, (s) => E.fold_(getEither(s), T.fail, T.succeed)));
  }
  set(a) {
    return this.use((value, _, setEither) => E.fold_(setEither(a), T.fail, value.set));
  }
}
class DerivedAll {
  constructor(use) {
    this.use = use;
    this._tag = "DerivedAll";
    this._typeId = TypeId;
    this.fold = this.fold.bind(this);
    this.foldAll = this.foldAll.bind(this);
    this.set = this.set.bind(this);
  }
  fold(ea, eb, ca, bd) {
    return this.use((value, getEither, setEither) => new DerivedAll((f) => f(value, (s) => E.fold_(getEither(s), (e) => E.left(eb(e)), bd), (c) => (s) => E.chain_(ca(c), (a) => E.fold_(setEither(a)(s), (e) => E.left(ea(e)), E.right)))));
  }
  foldAll(ea, eb, ec, ca, bd) {
    return this.use((value, getEither, setEither) => new DerivedAll((f) => f(value, (s) => E.fold_(getEither(s), (e) => E.left(eb(e)), bd), (c) => (s) => E.chain_(E.fold_(getEither(s), (e) => E.left(ec(e)), ca(c)), (a) => E.fold_(setEither(a)(s), (e) => E.left(ea(e)), E.right)))));
  }
  get get() {
    return this.use((value, getEither) => T.chain_(value.get, (a) => E.fold_(getEither(a), T.fail, T.succeed)));
  }
  set(a) {
    return this.use((value, _, setEither) => T.absolve(atomic.modify(value, (s) => E.fold_(setEither(a)(s), (e) => Tp.tuple(E.leftW(e), s), (s2) => Tp.tuple(E.right(void 0), s2)))));
  }
}
function concrete(self) {
  return self;
}
module.exports = __toCommonJS(XRef_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Atomic,
  Derived,
  DerivedAll,
  TypeId,
  concrete
});
